{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Instituto Tecnológico y de Estudios Superiores de Monterrey\n",
    "## Maestría en Inteligencia Artificial Aplicada\n",
    "### Proyecto Integrador (Gpo 10) - TC5035.10\n",
    "\n",
    "### **Proyecto: Diseño Acelerado de Fármacos**\n",
    "\n",
    "### Avance 2: Ingeniería de características\n",
    "\n",
    "#### **Docentes:**\n",
    "- Dra. Grettel Barceló Alonso - Profesor Titular\n",
    "- Dr. Luis Eduardo Falcón Morales - Profesor Titular\n",
    "- Dr. Horacio Martinez Alfaro  – Profesor Tutor\n",
    "\n",
    "#### **Miembros del equipo:**\n",
    " - Jose Luis Artigas Arroyo - A01794906\n",
    " - José Luis Fabela Portillo - A00748551\n",
    " - Alejandro Emmanuel Silva Ortega - A01794545"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importar librerias\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "import re\n",
    "from pfeature import pcp_wp\n",
    "from pfeature import ctd_wp\n",
    "from pfeature import sep_wp\n",
    "from pfeature import rri_wp\n",
    "from Bio import SeqIO\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sequence</th>\n",
       "      <td>(RRWQWR)2-K-(Ahx)</td>\n",
       "      <td>(RRWQWR)4-K2-(Ahx)2-C2</td>\n",
       "      <td>(XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Binding_Target</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Classification</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Activity</th>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>origen</th>\n",
       "      <td>DRAMP</td>\n",
       "      <td>DRAMP</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DCTPep</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRAMP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                0  \\\n",
       "Sequence                                        (RRWQWR)2-K-(Ahx)   \n",
       "Binding_Target                                                NaN   \n",
       "Classification                                                NaN   \n",
       "Activity        ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...   \n",
       "origen                                                      DRAMP   \n",
       "DCTPep                                                          0   \n",
       "DRAMP                                                           0   \n",
       "\n",
       "                                                                1  \\\n",
       "Sequence                                   (RRWQWR)4-K2-(Ahx)2-C2   \n",
       "Binding_Target                                                NaN   \n",
       "Classification                                                NaN   \n",
       "Activity        ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...   \n",
       "origen                                                      DRAMP   \n",
       "DCTPep                                                          0   \n",
       "DRAMP                                                           0   \n",
       "\n",
       "                                                                2  \n",
       "Sequence                           (XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ  \n",
       "Binding_Target                                      Not available  \n",
       "Classification  TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...  \n",
       "Activity                                ANTIMICROBIAL, ANTICANCER  \n",
       "origen                                             DCTPep & DRAMP  \n",
       "DCTPep                                                          0  \n",
       "DRAMP                                                           1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#leer csv con dataset preprocesado\n",
    "df_peptidos = pd.read_csv('../data/processed/peptide_processed2.csv')\n",
    "df_peptidos.head(3).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eliminacion de columnas redundantes\n",
    "Despues de un analis mas profundo, y una revision con el experto en el tema, se procedera a la eliminacion de columnas redundantes, ya que estos valores pueden ser obtenidos direcvtamente de la secuencia de peptidos, estas columnas son\n",
    "\n",
    "- Chiral: Los aminoácidos L/D consisten en péptidos.\n",
    "- Mass: Masa\n",
    "- Formula: Formula del péptido\n",
    "- Absent_amino_acids: Aminoacidos no incluidos\n",
    "- Common_amino_acids: Animoacido mas comun en la sequencia\n",
    "- pI: Punto isoeléctrico\n",
    "- Basic_residues: Número de residuos base\n",
    "- Acidic_residues: Número de residuos ácidos\n",
    "- Net_charge: Carga Neta\n",
    "- Polar_residues: Número de residuos polares\n",
    "- Hydrophobic_residues: Número de residuos hidrofóbicos\n",
    "- Hydrophobicity: Hidrofobicidad\n",
    "- Boman_Index: Índice de Boman\n",
    "- Half_Life: Vida media\n",
    "- Aliphatic_Index: Índice alifático\n",
    "- Extinction_Coefficient_cystines: Coeficiente de extinción de cistinas\n",
    "- Absorbance_280nm: Absorbencia a 280nm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop_cols = ['Chiral','Mass','Formula','Absent_amino_acids','Common_amino_acids','pI','Basic_residues','Acidic_residues','Net_charge','Polar_residues','Hydrophobic_residues','Hydrophobicity','Boman_Index',\n",
    "#              'Half_Life','Aliphatic_Index','Extinction_Coefficient_cystines','Absorbance_280nm']\n",
    "# df_peptidos.drop(drop_cols, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sequence</th>\n",
       "      <td>(RRWQWR)2-K-(Ahx)</td>\n",
       "      <td>(RRWQWR)4-K2-(Ahx)2-C2</td>\n",
       "      <td>(XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ</td>\n",
       "      <td>AAAAAAARRRRΦFKX</td>\n",
       "      <td>AAAAARRRRΦFKX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Binding_Target</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not available</td>\n",
       "      <td>Not available</td>\n",
       "      <td>Not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Classification</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Activity</th>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>origen</th>\n",
       "      <td>DRAMP</td>\n",
       "      <td>DRAMP</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DCTPep</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRAMP</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                0  \\\n",
       "Sequence                                        (RRWQWR)2-K-(Ahx)   \n",
       "Binding_Target                                                NaN   \n",
       "Classification                                                NaN   \n",
       "Activity        ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...   \n",
       "origen                                                      DRAMP   \n",
       "DCTPep                                                          0   \n",
       "DRAMP                                                           0   \n",
       "\n",
       "                                                                1  \\\n",
       "Sequence                                   (RRWQWR)4-K2-(Ahx)2-C2   \n",
       "Binding_Target                                                NaN   \n",
       "Classification                                                NaN   \n",
       "Activity        ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...   \n",
       "origen                                                      DRAMP   \n",
       "DCTPep                                                          0   \n",
       "DRAMP                                                           0   \n",
       "\n",
       "                                                                2  \\\n",
       "Sequence                           (XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ   \n",
       "Binding_Target                                      Not available   \n",
       "Classification  TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...   \n",
       "Activity                                ANTIMICROBIAL, ANTICANCER   \n",
       "origen                                             DCTPep & DRAMP   \n",
       "DCTPep                                                          0   \n",
       "DRAMP                                                           1   \n",
       "\n",
       "                                                                3  \\\n",
       "Sequence                                          AAAAAAARRRRΦFKX   \n",
       "Binding_Target                                      Not available   \n",
       "Classification  CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...   \n",
       "Activity                                ANTIMICROBIAL, ANTICANCER   \n",
       "origen                                             DCTPep & DRAMP   \n",
       "DCTPep                                                          1   \n",
       "DRAMP                                                           1   \n",
       "\n",
       "                                                                4  \n",
       "Sequence                                            AAAAARRRRΦFKX  \n",
       "Binding_Target                                      Not available  \n",
       "Classification  CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...  \n",
       "Activity                                ANTIMICROBIAL, ANTICANCER  \n",
       "origen                                             DCTPep & DRAMP  \n",
       "DCTPep                                                          1  \n",
       "DRAMP                                                           1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos.head(5).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analisis de secuencias de péptidos\n",
    "\n",
    "De acuerdo a la IUPAC[1], los aminoacidos se clasifican de acuerdo a la siguiente tabla\n",
    "\n",
    "\n",
    "|symbol|Amino acid|\n",
    "| :-:|:-:|\n",
    "|A\t| alanine|\n",
    "|B\t| aspartic acid or asparagine|\n",
    "|C\t| cysteine|\n",
    "|D\t| aspartic acid|\n",
    "|E\t| glutamic acid|\n",
    "|F\t| phenylalanine|\n",
    "|G\t| glycine|\n",
    "|H\t| histidine|\n",
    "|I\t| isoleucine|\n",
    "|K\t| lysine|\n",
    "|L\t| leucine|\n",
    "|M\t| methionine|\n",
    "|N\t| asparagine|\n",
    "|P\t| proline|\n",
    "|Q\t| glutamine|\n",
    "|R\t| arginine|\n",
    "|S\t| serine|\n",
    "|T\t| threonine|\n",
    "|U  | selenocysteine|\n",
    "|V\t| valine|\n",
    "|W\t| tryptophan|\n",
    "|X  | Desconocido u 'otro'|\n",
    "|Y\t| tyrosine|\n",
    "|Z\t| glutamic acid or glutamine|\n",
    "\n",
    "Por lo cual, se eliminaran de la secuencia los caractares diferentes a los indicados en esta lista, de igual forma se actualizara la longitud de la secuencia para hacerla coincidir con la longitud despues de eliminar los caracteres invalidos, se eliminaran los registros con cadenas que contengan \"-\" y numeros (1-9) ya que estas representan peptidos modificados y contienen compuestos quimicos los cuales modifican el péptido y estos no seran considerados para el modelo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Binding_Target</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Activity</th>\n",
       "      <th>origen</th>\n",
       "      <th>DCTPep</th>\n",
       "      <th>DRAMP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(RRWQWR)2-K-(Ahx)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>DRAMP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(RRWQWR)4-K2-(Ahx)2-C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...</td>\n",
       "      <td>DRAMP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ</td>\n",
       "      <td>Not available</td>\n",
       "      <td>TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAARRRRΦFKX</td>\n",
       "      <td>Not available</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAARRRRΦFKX</td>\n",
       "      <td>Not available</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12724</th>\n",
       "      <td>yAIIXYNKYXNC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANTIMICROBIAL,ANTIVIRAL</td>\n",
       "      <td>DRAMP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12725</th>\n",
       "      <td>yf</td>\n",
       "      <td>Not available</td>\n",
       "      <td>TUMOR ACTIVE PEPTIDE</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12726</th>\n",
       "      <td>ykqchkkGGkkGsG</td>\n",
       "      <td>Not available</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12727</th>\n",
       "      <td>zV</td>\n",
       "      <td>Not available</td>\n",
       "      <td>TUMOR ACTIVE PEPTIDE</td>\n",
       "      <td>ANTIMICROBIAL, ANTICANCER</td>\n",
       "      <td>DCTPep &amp; DRAMP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12728</th>\n",
       "      <td>ⓍKAPKVVⓍLSKALEYL</td>\n",
       "      <td>Not available</td>\n",
       "      <td>CANCER TARGETED PEPTIDES##STAPLED PEPTIDES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DCTPep</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12729 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sequence Binding_Target  \\\n",
       "0                   (RRWQWR)2-K-(Ahx)            NaN   \n",
       "1              (RRWQWR)4-K2-(Ahx)2-C2            NaN   \n",
       "2      (XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ  Not available   \n",
       "3                     AAAAAAARRRRΦFKX  Not available   \n",
       "4                       AAAAARRRRΦFKX  Not available   \n",
       "...                               ...            ...   \n",
       "12724                    yAIIXYNKYXNC            NaN   \n",
       "12725                              yf  Not available   \n",
       "12726                  ykqchkkGGkkGsG  Not available   \n",
       "12727                              zV  Not available   \n",
       "12728                ⓍKAPKVVⓍLSKALEYL  Not available   \n",
       "\n",
       "                                          Classification  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2      TARGETED PEPTIDE CONJUGATES##MEMBRANE-TARGETED...   \n",
       "3      CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...   \n",
       "4      CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...   \n",
       "...                                                  ...   \n",
       "12724                                                NaN   \n",
       "12725                               TUMOR ACTIVE PEPTIDE   \n",
       "12726  CANCER TARGETED PEPTIDES##MEMBRANE-TARGETED##C...   \n",
       "12727                               TUMOR ACTIVE PEPTIDE   \n",
       "12728         CANCER TARGETED PEPTIDES##STAPLED PEPTIDES   \n",
       "\n",
       "                                                Activity          origen  \\\n",
       "0      ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...           DRAMP   \n",
       "1      ANTIMICROBIAL, ANTIBACTERIAL, ANTI-GRAM+, ANTI...           DRAMP   \n",
       "2                              ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "3                              ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "4                              ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "...                                                  ...             ...   \n",
       "12724                            ANTIMICROBIAL,ANTIVIRAL           DRAMP   \n",
       "12725                          ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "12726                          ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "12727                          ANTIMICROBIAL, ANTICANCER  DCTPep & DRAMP   \n",
       "12728                                                NaN          DCTPep   \n",
       "\n",
       "       DCTPep  DRAMP  \n",
       "0           0      0  \n",
       "1           0      0  \n",
       "2           0      1  \n",
       "3           1      1  \n",
       "4           1      1  \n",
       "...       ...    ...  \n",
       "12724       0      0  \n",
       "12725       1      1  \n",
       "12726       1      1  \n",
       "12727       1      1  \n",
       "12728       1      0  \n",
       "\n",
       "[12729 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos#[\"Sequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                     (RRWQWR)2-K-(Ahx)\n",
       "1                (RRWQWR)4-K2-(Ahx)2-C2\n",
       "2        (XGLFGC)(XGLFGC)KGRKKRRQRRRPPQ\n",
       "3                       AAAAAAARRRRΦFKX\n",
       "4                         AAAAARRRRΦFKX\n",
       "                      ...              \n",
       "12724                      yAIIXYNKYXNC\n",
       "12725                                yf\n",
       "12726                    ykqchkkGGkkGsG\n",
       "12727                                zV\n",
       "12728                  ⓍKAPKVVⓍLSKALEYL\n",
       "Name: Sequence, Length: 12729, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos[\"Sequence\"].map(lambda x: re.sub(r\"[\\\"]\",'',x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "      <th>64</th>\n",
       "      <th>65</th>\n",
       "      <th>66</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>-</td>\n",
       "      <td>(</td>\n",
       "      <td>h</td>\n",
       "      <td>W</td>\n",
       "      <td>A</td>\n",
       "      <td>x</td>\n",
       "      <td>)</td>\n",
       "      <td>...</td>\n",
       "      <td>q</td>\n",
       "      <td>d</td>\n",
       "      <td>工</td>\n",
       "      <td>u</td>\n",
       "      <td>b</td>\n",
       "      <td>j</td>\n",
       "      <td>]</td>\n",
       "      <td>[</td>\n",
       "      <td>g</td>\n",
       "      <td>z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>3</td>\n",
       "      <td>9071</td>\n",
       "      <td>7214</td>\n",
       "      <td>26</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>5002</td>\n",
       "      <td>7895</td>\n",
       "      <td>87</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>0.02</td>\n",
       "      <td>71.26</td>\n",
       "      <td>56.67</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.09</td>\n",
       "      <td>39.3</td>\n",
       "      <td>62.02</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0      1      2    3     4     5     6      7     8     9   ...  \\\n",
       "Character     2      K      R    -     (     h     W      A     x     )  ...   \n",
       "Count         3   9071   7214   26     6    11  5002   7895    87     6  ...   \n",
       "%          0.02  71.26  56.67  0.2  0.05  0.09  39.3  62.02  0.68  0.05  ...   \n",
       "\n",
       "             57    58    59    60    61    62    63    64    65    66  \n",
       "Character     q     d     工     u     b     j     ]     [     g     z  \n",
       "Count        21     8     1     1     1     3     1     1    29     1  \n",
       "%          0.16  0.06  0.01  0.01  0.01  0.02  0.01  0.01  0.23  0.01  \n",
       "\n",
       "[3 rows x 67 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Crear un diccionario para contar la aparición de cada carácter de la secuencia de aminoácidos\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in df_peptidos['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / df_peptidos.shape[0]).round(4) *100\n",
    "\n",
    "# Número de péptidos por caracter\n",
    "char_count_df.T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Es importante identificar los caracteres contenidos en la secuencia de aminoácidos de los péptidos, por tal motivo vamos a analizar su ocurrencia en el total de péptidos. La lista utilizada es la que se describio anteriormente (IUPAC), sin embargo, para obtener las características con la librería de Pfeatures no se reconocen 4 de estos: B, U, X, Z. Asímismo existen otros caracteres no incluidos en IUPAC por lo que es necesario analizar que tan frecuentes son."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de aminoácidos IUPAC: 24\n",
      "Total de aminoácidos Pfeatures: 20\n",
      "Total de aminoácidos que están en UIPAC pero no reconoce Pfeatures: 4\n"
     ]
    }
   ],
   "source": [
    "# Aminoácidos Pfeature\n",
    "aa_pfeature = ['A','C','D','E','F','G','H','I','K','L','M','N','P',\n",
    "              'Q','R','S','T','V','W','Y']\n",
    "\n",
    "# Aminoácidos IUPAC - Pfeatures\n",
    "aa_iupac = ['B','U','X','Z']\n",
    "\n",
    "# Aminoácidos en minúsculas\n",
    "aa_min = [elemento.lower() for elemento in (aa_pfeature + aa_iupac)]\n",
    "\n",
    "print ('Total de aminoácidos IUPAC:', len(aa_iupac) + len(aa_pfeature))\n",
    "print ('Total de aminoácidos Pfeatures:', len(aa_pfeature))\n",
    "print ('Total de aminoácidos que están en UIPAC pero no reconoce Pfeatures:', len(aa_iupac))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres únicos:  67\n"
     ]
    }
   ],
   "source": [
    "# Crear un diccionario para contar la aparición de cada carácter\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in df_peptidos['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / df_peptidos.shape[0]).round(4) *100\n",
    "\n",
    "\n",
    "print('Total de caracteres únicos: ', len(char_count_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>13</th>\n",
       "      <th>1</th>\n",
       "      <th>27</th>\n",
       "      <th>16</th>\n",
       "      <th>7</th>\n",
       "      <th>19</th>\n",
       "      <th>2</th>\n",
       "      <th>15</th>\n",
       "      <th>24</th>\n",
       "      <th>14</th>\n",
       "      <th>25</th>\n",
       "      <th>22</th>\n",
       "      <th>20</th>\n",
       "      <th>23</th>\n",
       "      <th>6</th>\n",
       "      <th>10</th>\n",
       "      <th>21</th>\n",
       "      <th>11</th>\n",
       "      <th>28</th>\n",
       "      <th>26</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>L</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "      <td>F</td>\n",
       "      <td>S</td>\n",
       "      <td>P</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>W</td>\n",
       "      <td>Q</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>9541</td>\n",
       "      <td>9071</td>\n",
       "      <td>8101</td>\n",
       "      <td>7981</td>\n",
       "      <td>7895</td>\n",
       "      <td>7730</td>\n",
       "      <td>7214</td>\n",
       "      <td>7112</td>\n",
       "      <td>6597</td>\n",
       "      <td>5976</td>\n",
       "      <td>5772</td>\n",
       "      <td>5569</td>\n",
       "      <td>5054</td>\n",
       "      <td>5019</td>\n",
       "      <td>5002</td>\n",
       "      <td>4846</td>\n",
       "      <td>4813</td>\n",
       "      <td>4199</td>\n",
       "      <td>3387</td>\n",
       "      <td>3193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>74.95</td>\n",
       "      <td>71.26</td>\n",
       "      <td>63.64</td>\n",
       "      <td>62.7</td>\n",
       "      <td>62.02</td>\n",
       "      <td>60.73</td>\n",
       "      <td>56.67</td>\n",
       "      <td>55.87</td>\n",
       "      <td>51.83</td>\n",
       "      <td>46.95</td>\n",
       "      <td>45.35</td>\n",
       "      <td>43.75</td>\n",
       "      <td>39.7</td>\n",
       "      <td>39.43</td>\n",
       "      <td>39.3</td>\n",
       "      <td>38.07</td>\n",
       "      <td>37.81</td>\n",
       "      <td>32.99</td>\n",
       "      <td>26.61</td>\n",
       "      <td>25.08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              13     1      27    16     7      19     2      15     24  \\\n",
       "Character      L      K      I     G      A      V      R      F      S   \n",
       "Count       9541   9071   8101  7981   7895   7730   7214   7112   6597   \n",
       "%          74.95  71.26  63.64  62.7  62.02  60.73  56.67  55.87  51.83   \n",
       "\n",
       "              14     25     22    20     23    6      10     21     11     28  \\\n",
       "Character      P      N      T     E      D     W      Q      Y      C      H   \n",
       "Count       5976   5772   5569  5054   5019  5002   4846   4813   4199   3387   \n",
       "%          46.95  45.35  43.75  39.7  39.43  39.3  38.07  37.81  32.99  26.61   \n",
       "\n",
       "              26  \n",
       "Character      M  \n",
       "Count       3193  \n",
       "%          25.08  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos Pfeatures Mayúsculas\n",
    "char_count_aapf = char_count_df[char_count_df['Character'].isin(aa_pfeature)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aapf))\n",
    "char_count_aapf.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>17</th>\n",
       "      <th>50</th>\n",
       "      <th>29</th>\n",
       "      <th>55</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>X</td>\n",
       "      <td>Z</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>918</td>\n",
       "      <td>371</td>\n",
       "      <td>54</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>7.21</td>\n",
       "      <td>2.91</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             17    50    29    55\n",
       "Character     X     Z     B     U\n",
       "Count       918   371    54     6\n",
       "%          7.21  2.91  0.42  0.05"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos UIPAC no incluídos en Pfeatures Mayúsculas\n",
    "char_count_aauipac = char_count_df[char_count_df['Character'].isin(aa_iupac)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aauipac))\n",
    "char_count_aauipac.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La ocurrencia de Z, B, U es muy baja por lo que se procederá a eliminar todos los péptidos que las contienen. \n",
    "Para el caso de los péptidos con 'x' se re-escribiran aquellas que solo tengan X al inicio o al final. Por ejemplo, si la secuencia es XXVKKGLISY reescribirla como VKKGLISY, si la secuencia es KKGLISYXXX entonces se reescribirá como KKGLISY. Sin embargo, si la secuencia es KKGXXLISY entonces se eliminará la secuencia del dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>30</th>\n",
       "      <th>35</th>\n",
       "      <th>34</th>\n",
       "      <th>32</th>\n",
       "      <th>38</th>\n",
       "      <th>8</th>\n",
       "      <th>44</th>\n",
       "      <th>43</th>\n",
       "      <th>41</th>\n",
       "      <th>37</th>\n",
       "      <th>...</th>\n",
       "      <th>42</th>\n",
       "      <th>56</th>\n",
       "      <th>39</th>\n",
       "      <th>57</th>\n",
       "      <th>51</th>\n",
       "      <th>5</th>\n",
       "      <th>58</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>66</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>l</td>\n",
       "      <td>r</td>\n",
       "      <td>k</td>\n",
       "      <td>f</td>\n",
       "      <td>i</td>\n",
       "      <td>x</td>\n",
       "      <td>a</td>\n",
       "      <td>w</td>\n",
       "      <td>v</td>\n",
       "      <td>n</td>\n",
       "      <td>...</td>\n",
       "      <td>y</td>\n",
       "      <td>c</td>\n",
       "      <td>e</td>\n",
       "      <td>q</td>\n",
       "      <td>m</td>\n",
       "      <td>h</td>\n",
       "      <td>d</td>\n",
       "      <td>u</td>\n",
       "      <td>b</td>\n",
       "      <td>z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>180</td>\n",
       "      <td>179</td>\n",
       "      <td>156</td>\n",
       "      <td>103</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>77</td>\n",
       "      <td>76</td>\n",
       "      <td>61</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>1.41</td>\n",
       "      <td>1.41</td>\n",
       "      <td>1.23</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.42</td>\n",
       "      <td>...</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             30    35    34    32    38    8    44   43    41    37  ...  \\\n",
       "Character     l     r     k     f     i     x    a    w     v     n  ...   \n",
       "Count       180   179   156   103    87    87   77   76    61    54  ...   \n",
       "%          1.41  1.41  1.23  0.81  0.68  0.68  0.6  0.6  0.48  0.42  ...   \n",
       "\n",
       "             42    56    39    57    51    5     58    60    61    66  \n",
       "Character     y     c     e     q     m     h     d     u     b     z  \n",
       "Count        28    28    21    21    14    11     8     1     1     1  \n",
       "%          0.22  0.22  0.16  0.16  0.11  0.09  0.06  0.01  0.01  0.01  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos UIPAC no incluídos en Pfeatures Mayúsculas\n",
    "char_count_aamin = char_count_df[char_count_df['Character'].isin(aa_min)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aamin))\n",
    "char_count_aamin.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La distincion entre los tipos de aminoácidos L o sinteticos y D o naturales se puede ver en estos resultados, siendo representados los tipo D con letras mayusculuas y los tipo L con minusculas. Gracias a este analisis podemos mostrar que no es relevante mantener la distincion entre estos aminoácidos y se hara un tratamiento a estos datos para convertir los caracteres en minusculas a mayusculas mas adelante.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres no incluidos en el UIPAC:  19\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>33</th>\n",
       "      <th>3</th>\n",
       "      <th>45</th>\n",
       "      <th>31</th>\n",
       "      <th>18</th>\n",
       "      <th>4</th>\n",
       "      <th>9</th>\n",
       "      <th>40</th>\n",
       "      <th>62</th>\n",
       "      <th>0</th>\n",
       "      <th>52</th>\n",
       "      <th>54</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>48</th>\n",
       "      <th>59</th>\n",
       "      <th>12</th>\n",
       "      <th>63</th>\n",
       "      <th>64</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>Ⓧ</td>\n",
       "      <td>-</td>\n",
       "      <td>,</td>\n",
       "      <td>O</td>\n",
       "      <td>Φ</td>\n",
       "      <td>(</td>\n",
       "      <td>)</td>\n",
       "      <td></td>\n",
       "      <td>j</td>\n",
       "      <td>2</td>\n",
       "      <td>J</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>工</td>\n",
       "      <td>4</td>\n",
       "      <td>]</td>\n",
       "      <td>[</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>236</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>1.85</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             33   3     45    31    18    4     9     40    62    0     52  \\\n",
       "Character     Ⓧ    -     ,     O     Φ     (     )           j     2     J   \n",
       "Count       236   26    18    18    11     6     6     5     3     3     2   \n",
       "%          1.85  0.2  0.14  0.14  0.09  0.05  0.05  0.04  0.02  0.02  0.02   \n",
       "\n",
       "             54    46    47    48    59    12    63    64  \n",
       "Character           6     1     7     工     4     ]     [  \n",
       "Count         2     1     1     1     1     1     1     1  \n",
       "%          0.02  0.01  0.01  0.01  0.01  0.01  0.01  0.01  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Número de péptidos en los que está cada caracter que no es aminoácido IUPAC:\n",
    "char_count_otros = char_count_df[~char_count_df['Character'].isin(aa_pfeature + aa_iupac + aa_min)]\n",
    "print('Total de caracteres no incluidos en el UIPAC: ', len(char_count_otros))\n",
    "char_count_otros.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['2', '-', '(', ')', '4', 'Φ', 'O', 'Ⓧ', ' ', ',', '6', '1', '7',\n",
       "       'J', '\\xa0', '工', 'j', ']', '['], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "char_count_otros['Character'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La frecuencia de los caractéres no incluidos en IUPAC es muy baja, por lo que también se eliminarán los péptidos que las contienen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ajuste de las secuencias de aminoácidos\n",
    "filtered_df = df_peptidos.copy()\n",
    "# 1. Poner en mayúsculas todos los aminoácidos de las secuencias\n",
    "filtered_df['Sequence'] = filtered_df['Sequence'].str.upper()\n",
    "# 2. Identificar los péptidos con valores de X\n",
    "aa_x = filtered_df[filtered_df['Sequence'].apply(lambda x: 'X' in x)]\n",
    "# 3. Conservar solo los péptidos que contengan los aminoácidos para pfeatures\n",
    "filtered_df = filtered_df[filtered_df['Sequence'].apply(lambda x: all(char in aa_pfeature for char in x))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de péptidos con X 960\n",
      "Péptidos que empiezan con X:  280\n",
      "Péptidos que finalizan con X:  232\n",
      "Péptidos con X intermedia: 729\n",
      "Peptidos conservados:  231\n"
     ]
    }
   ],
   "source": [
    "# Identifica todos péptidos con valores de X\n",
    "total_x = len(aa_x)\n",
    "print('Total de péptidos con X', total_x)\n",
    "\n",
    "# Inician con X \n",
    "inicio_x = aa_x[aa_x['Sequence'].str.startswith('X')]\n",
    "print('Péptidos que empiezan con X: ', len(inicio_x))\n",
    "# Función para quitar los caracteres 'x' al inicio de la secuencia\n",
    "def quitar_x_inicio(secuencia):\n",
    "    return re.sub(r'^X+', '', secuencia)\n",
    "# Aplicar la función a la columna 'Sequence'\n",
    "aa_x['Sequence'] = aa_x['Sequence'].apply(quitar_x_inicio)\n",
    "\n",
    "# Finalizan con X\n",
    "fin_x = aa_x[aa_x['Sequence'].str.endswith('X')].copy()\n",
    "print('Péptidos que finalizan con X: ', len(fin_x))\n",
    "# Función para quitar los caracteres 'x' al final de la secuencia\n",
    "def quitar_x_final(secuencia):\n",
    "    return re.sub(r'X+$', '', secuencia)\n",
    "# Aplicar la función a la columna 'Sequence'\n",
    "aa_x['Sequence'] = aa_x['Sequence'].apply(quitar_x_final)\n",
    "\n",
    "# Elimina los péptidos con X intermedia\n",
    "aa_x = aa_x[aa_x['Sequence'].apply(lambda x: all(char in aa_pfeature for char in x))]\n",
    "print('Péptidos con X intermedia:', total_x - len(aa_x))\n",
    "print('Peptidos conservados: ', len(aa_x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de péptidos:  12729\n",
      "Péptidos solo con aminoácidos para pFeatures:  11318\n",
      "Péptidos eliminados:  11.08 %\n"
     ]
    }
   ],
   "source": [
    "# Agrega los peptidos con X modificados al dataframe de péptidos depurados\n",
    "filtered_df = pd.concat([filtered_df, aa_x], ignore_index=True)\n",
    "\n",
    "print('Total de péptidos: ', df_peptidos.shape[0])\n",
    "print('Péptidos solo con aminoácidos para pFeatures: ', filtered_df.shape[0])\n",
    "print('Péptidos eliminados: ', round((df_peptidos.shape[0]-filtered_df.shape[0])/df_peptidos.shape[0],4) * 100, '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres únicos:  20\n"
     ]
    }
   ],
   "source": [
    "# Valida los aminoácidos contenidos\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in filtered_df['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / filtered_df.shape[0]).round(4) *100\n",
    "\n",
    "\n",
    "print('Total de caracteres únicos: ', len(char_count_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        0\n",
       "1        1\n",
       "2        0\n",
       "3        1\n",
       "4        0\n",
       "        ..\n",
       "11313    1\n",
       "11314    1\n",
       "11315    1\n",
       "11316    1\n",
       "11317    1\n",
       "Name: DCTPep, Length: 11318, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df['DCTPep']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df.to_csv('../data/processed/peptide_filtered.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convertir dataframe to FASTA \n",
    "\n",
    "Para los siguientes pasos del proceso, es necesario hacer la conversion de los datos filtrados a formato FASTA, este formato es el estandard usado para las secuencias de aminoácidos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('../data/processed/DCTPep.fasta','w', encoding='UTF8') as f:\n",
    "    \n",
    "    for _, row in filtered_df.iterrows():\n",
    "        f.write(f\">{row['DCTPep']}\\n{row['Sequence']}\\n\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ejecutar CD-HIT\n",
    "\n",
    "CD-HIT es un programa ampliamente utilizado para agrupar secuencias biológicas con el fin de reducir la redundancia de secuencias y mejorar el rendimiento de otros análisis de secuencias. En respuesta al rápido aumento de la cantidad de datos de secuenciación producidos por las tecnologías de secuenciación de próxima generación, hemos desarrollado un nuevo programa CD-HIT acelerado con una novedosa estrategia de paralelización y algunas otras técnicas para permitir la agrupación eficiente de dichos conjuntos de datos. Nuestras pruebas demostraron una aceleración muy buena derivada de la paralelización para hasta ∼24 núcleos y una aceleración cuasi-lineal para hasta ∼8 núcleos. El CD-HIT mejorado es capaz de manejar conjuntos de datos muy grandes en mucho menos tiempo que las versiones anteriores[2]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Aug 20 2021, 08:39:56\n",
      "Command: cd-hit -i ../data/processed/DCTPep.fasta -o\n",
      "         ../data/processed/DCTPep.txt -c 0.99\n",
      "\n",
      "Started: Thu Feb  6 10:32:39 2025\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 9438\n",
      "longest and shortest : 101 and 11\n",
      "Total letters: 265896\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 1M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 77M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 90300214\n",
      "\n",
      "\n",
      "comparing sequences from          0  to       9438\n",
      ".........\n",
      "     9438  finished       8091  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 79M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.13\n",
      "\n",
      "STDERR: w\u0000s\u0000l\u0000:\u0000 \u0000A\u0000 \u0000l\u0000o\u0000c\u0000a\u0000l\u0000h\u0000o\u0000s\u0000t\u0000 \u0000p\u0000r\u0000o\u0000x\u0000y\u0000 \u0000c\u0000o\u0000n\u0000f\u0000i\u0000g\u0000u\u0000r\u0000a\u0000t\u0000i\u0000o\u0000n\u0000 \u0000w\u0000a\u0000s\u0000 \u0000d\u0000e\u0000t\u0000e\u0000c\u0000t\u0000e\u0000d\u0000 \u0000b\u0000u\u0000t\u0000 \u0000n\u0000o\u0000t\u0000 \u0000m\u0000i\u0000r\u0000r\u0000o\u0000r\u0000e\u0000d\u0000 \u0000i\u0000n\u0000t\u0000o\u0000 \u0000W\u0000S\u0000L\u0000.\u0000 \u0000W\u0000S\u0000L\u0000 \u0000i\u0000n\u0000 \u0000N\u0000A\u0000T\u0000 \u0000m\u0000o\u0000d\u0000e\u0000 \u0000d\u0000o\u0000e\u0000s\u0000 \u0000n\u0000o\u0000t\u0000 \u0000s\u0000u\u0000p\u0000p\u0000o\u0000r\u0000t\u0000 \u0000l\u0000o\u0000c\u0000a\u0000l\u0000h\u0000o\u0000s\u0000t\u0000 \u0000p\u0000r\u0000o\u0000x\u0000i\u0000e\u0000s\u0000.\u0000\n",
      "\u0000\n",
      "\u0000\n"
     ]
    }
   ],
   "source": [
    "#Ejecutar cd-hit en colab, para windows usar el comando con wsl\n",
    "#! cd-hit -i data/processed/DCTPep.fasta -o data/processed/DCTPep.txt -c 0.99\n",
    "\n",
    "import subprocess\n",
    "\n",
    "# Define the WSL command buscar \n",
    "command = [\"wsl\", \"cd-hit\", \"-i\", \"../data/processed/DCTPep.fasta\", \"-o\", \"../data/processed/DCTPep.txt\", \"-c\", \"0.99\"]\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(command, capture_output=True, text=True)\n",
    "\n",
    "# Print the output and errors\n",
    "print(\"STDOUT:\", result.stdout)\n",
    "print(\"STDERR:\", result.stderr)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Funciones de Pfeature\n",
    "\n",
    "La librería pfeature es una herramienta de Python utilizada para extraer y analizar características de secuencias proteicas y nucleotídicas. Proporciona métodos para calcular descriptores físicos, químicos, estadísticos y estructurales de biomoléculas, lo que la hace útil en bioinformática, predicción de propiedades funcionales y clasificación de secuencias en aprendizaje automático.\n",
    "\n",
    "\n",
    "## Aplicaciones en Bioinformática\n",
    "pfeature se usa ampliamente en:\n",
    "- Predicción de propiedades de proteínas (hidrofobicidad, accesibilidad al solvente, estabilidad).\n",
    "- Clasificación de proteínas en modelos de Machine Learning.\n",
    "- Análisis de secuencias en estudios evolutivos.\n",
    "- Desarrollo de modelos para predicción de interacciones proteína-ligando.\n",
    "\n",
    "pfeature es una poderosa herramienta en bioinformática para analizar secuencias biológicas de manera rápida y eficiente. Su capacidad de generar múltiples descriptores permite su aplicación en aprendizaje automático, descubrimiento de fármacos y análisis estructural de proteínas[3].\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lee el archivo creado por CD-HIT y almacena las etiquetas en un archivo CSV para su uso posterior\n",
    "cd_hit_path = \"../data/processed/DCTPep.txt\"\n",
    "records_cd_hit = list(SeqIO.parse(cd_hit_path, \"fasta\"))\n",
    "seqs = []\n",
    "for records in records_cd_hit:\n",
    "    seqs.append(re.sub(r'[Seq\\(\\'][\\']','',str(records.seq)))\n",
    "\n",
    "with open('../data/processed/DTCPep_y.csv','w', encoding='UTF8') as f:\n",
    "    f.write('Label\\n')\n",
    "    for record in records_cd_hit:\n",
    "        f.write(f'{record.id} \\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Physico-Chemical Properties Composition: 30 Descriptores\n",
    "\n",
    "Este modulo calcula la fracción de cada propiedas físico-química de una secuencia. Las siguientes propiedades se han incorporado a Pfeature para calcular caracteristicas composicionales basadas en propiedades físico-químicas estandard.\n",
    "\n",
    "- Positively Charged → Carga Positiva\n",
    "- Negatively Charged → Carga Negativa\n",
    "- Neutral Charge → Carga Neutra\n",
    "- Polarity in residues → Polaridad en residuos\n",
    "- Non-polarity in residues → No polaridad en residuos\n",
    "- Aliphaticity → Alifaticidad\n",
    "- Cyclicity → Ciclicidad\n",
    "- Aromaticity → Aromaticidad\n",
    "- Acidity → Acidez\n",
    "- Basicity → Basicidad\n",
    "- Neutral (pH) → Neutro (pH)\n",
    "- Hydrophobicity → Hidrofobicidad\n",
    "- Hydrophilicity → Hidrofilicidad\n",
    "- Neutral towards water → Neutro frente al agua\n",
    "- Hydroxylic → Hidroxílico\n",
    "- Sulphur Content → Contenido de Azufre\n",
    "- Tiny → Pequeño\n",
    "- Small → Mediano\n",
    "- Large → Grande\n",
    "\n",
    "parfa el cálculo de estas propiedades se usa la siguiente formula \n",
    "\n",
    "### 𝐶𝑃𝑖 = 𝑃𝑖/𝐿 \n",
    "\n",
    "donde\n",
    "\n",
    "- 𝐶𝑃𝑖 = Propiedades físico-quimicas de la composicion de tipo 𝑖.\n",
    "- 𝑃𝑖 = suma de la propiedad de tipo 𝑖.\n",
    "- 𝐿 = Longitud de la secuencia. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCP_PC</th>\n",
       "      <th>PCP_NC</th>\n",
       "      <th>PCP_NE</th>\n",
       "      <th>PCP_PO</th>\n",
       "      <th>PCP_NP</th>\n",
       "      <th>PCP_AL</th>\n",
       "      <th>PCP_CY</th>\n",
       "      <th>PCP_AR</th>\n",
       "      <th>PCP_AC</th>\n",
       "      <th>PCP_BS</th>\n",
       "      <th>...</th>\n",
       "      <th>PCP_SA_EX</th>\n",
       "      <th>PCP_SA_IN</th>\n",
       "      <th>PCP_TN</th>\n",
       "      <th>PCP_SM</th>\n",
       "      <th>PCP_LR</th>\n",
       "      <th>PCP_Z1</th>\n",
       "      <th>PCP_Z2</th>\n",
       "      <th>PCP_Z3</th>\n",
       "      <th>PCP_Z4</th>\n",
       "      <th>PCP_Z5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.093</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.837</td>\n",
       "      <td>0.302</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209</td>\n",
       "      <td>0.186</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.721</td>\n",
       "      <td>0.279</td>\n",
       "      <td>0.799</td>\n",
       "      <td>-0.798</td>\n",
       "      <td>0.434</td>\n",
       "      <td>-0.325</td>\n",
       "      <td>-0.490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.124</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.794</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.443</td>\n",
       "      <td>0.371</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.124</td>\n",
       "      <td>...</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.227</td>\n",
       "      <td>0.268</td>\n",
       "      <td>0.691</td>\n",
       "      <td>0.309</td>\n",
       "      <td>0.618</td>\n",
       "      <td>-0.324</td>\n",
       "      <td>0.238</td>\n",
       "      <td>-0.341</td>\n",
       "      <td>-0.152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.107</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.357</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-0.263</td>\n",
       "      <td>-0.330</td>\n",
       "      <td>-0.695</td>\n",
       "      <td>-0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.250</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.300</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.400</td>\n",
       "      <td>1.129</td>\n",
       "      <td>-0.597</td>\n",
       "      <td>0.222</td>\n",
       "      <td>-0.022</td>\n",
       "      <td>-0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.150</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.150</td>\n",
       "      <td>...</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.650</td>\n",
       "      <td>0.962</td>\n",
       "      <td>-0.002</td>\n",
       "      <td>-0.257</td>\n",
       "      <td>-0.820</td>\n",
       "      <td>0.281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>0.037</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.926</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.704</td>\n",
       "      <td>0.296</td>\n",
       "      <td>-0.329</td>\n",
       "      <td>-1.297</td>\n",
       "      <td>0.556</td>\n",
       "      <td>-0.407</td>\n",
       "      <td>-0.425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8087</th>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.333</td>\n",
       "      <td>1.373</td>\n",
       "      <td>-0.309</td>\n",
       "      <td>0.556</td>\n",
       "      <td>-0.242</td>\n",
       "      <td>-0.268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8088</th>\n",
       "      <td>0.260</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0.380</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.340</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.480</td>\n",
       "      <td>0.887</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>0.129</td>\n",
       "      <td>-0.064</td>\n",
       "      <td>-0.445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8089</th>\n",
       "      <td>0.143</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.857</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.143</td>\n",
       "      <td>...</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.607</td>\n",
       "      <td>0.393</td>\n",
       "      <td>-0.636</td>\n",
       "      <td>-0.045</td>\n",
       "      <td>0.224</td>\n",
       "      <td>0.318</td>\n",
       "      <td>0.674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8090</th>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.857</td>\n",
       "      <td>1.159</td>\n",
       "      <td>0.873</td>\n",
       "      <td>-0.456</td>\n",
       "      <td>-0.630</td>\n",
       "      <td>-0.307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8091 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PCP_PC  PCP_NC  PCP_NE  PCP_PO  PCP_NP  PCP_AL  PCP_CY  PCP_AR  PCP_AC  \\\n",
       "0      0.093   0.070   0.837   0.302   0.442   0.372   0.000   0.140   0.070   \n",
       "1      0.124   0.082   0.794   0.216   0.443   0.371   0.082   0.093   0.082   \n",
       "2      0.107   0.179   0.714   0.214   0.429   0.393   0.000   0.107   0.179   \n",
       "3      0.250   0.075   0.675   0.275   0.350   0.300   0.000   0.050   0.075   \n",
       "4      0.150   0.317   0.533   0.150   0.367   0.283   0.033   0.067   0.317   \n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "8086   0.037   0.037   0.926   0.444   0.481   0.481   0.111   0.074   0.037   \n",
       "8087   0.190   0.095   0.714   0.381   0.238   0.190   0.095   0.048   0.095   \n",
       "8088   0.260   0.080   0.660   0.380   0.240   0.220   0.020   0.140   0.080   \n",
       "8089   0.143   0.000   0.857   0.107   0.750   0.643   0.393   0.143   0.000   \n",
       "8090   0.286   0.286   0.429   0.429   0.000   0.000   0.000   0.286   0.286   \n",
       "\n",
       "      PCP_BS  ...  PCP_SA_EX  PCP_SA_IN  PCP_TN  PCP_SM  PCP_LR  PCP_Z1  \\\n",
       "0      0.093  ...      0.209      0.186   0.488   0.721   0.279   0.799   \n",
       "1      0.124  ...      0.330      0.227   0.268   0.691   0.309   0.618   \n",
       "2      0.107  ...      0.429      0.107   0.214   0.357   0.643   0.070   \n",
       "3      0.250  ...      0.250      0.175   0.450   0.600   0.400   1.129   \n",
       "4      0.150  ...      0.533      0.150   0.167   0.350   0.650   0.962   \n",
       "...      ...  ...        ...        ...     ...     ...     ...     ...   \n",
       "8086   0.037  ...      0.148      0.333   0.407   0.704   0.296  -0.329   \n",
       "8087   0.190  ...      0.381      0.381   0.381   0.667   0.333   1.373   \n",
       "8088   0.260  ...      0.400      0.220   0.340   0.520   0.480   0.887   \n",
       "8089   0.143  ...      0.429      0.464   0.179   0.607   0.393  -0.636   \n",
       "8090   0.286  ...      0.571      0.429   0.143   0.143   0.857   1.159   \n",
       "\n",
       "      PCP_Z2  PCP_Z3  PCP_Z4  PCP_Z5  \n",
       "0     -0.798   0.434  -0.325  -0.490  \n",
       "1     -0.324   0.238  -0.341  -0.152  \n",
       "2     -0.263  -0.330  -0.695  -0.012  \n",
       "3     -0.597   0.222  -0.022  -0.250  \n",
       "4     -0.002  -0.257  -0.820   0.281  \n",
       "...      ...     ...     ...     ...  \n",
       "8086  -1.297   0.556  -0.407  -0.425  \n",
       "8087  -0.309   0.556  -0.242  -0.268  \n",
       "8088  -0.247   0.129  -0.064  -0.445  \n",
       "8089  -0.045   0.224   0.318   0.674  \n",
       "8090   0.873  -0.456  -0.630  -0.307  \n",
       "\n",
       "[8091 rows x 30 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Physico-Chemical Properties Composition (PCP)\n",
    "\n",
    "def pcp(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'dpc.csv'\n",
    "  df_out = pcp_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_pcp = pcp(cd_hit_path)\n",
    "feature_pcp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlacion: Descarta características redundantes con alta correlación entre sí."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCP_PC</th>\n",
       "      <th>PCP_NC</th>\n",
       "      <th>PCP_NE</th>\n",
       "      <th>PCP_PO</th>\n",
       "      <th>PCP_NP</th>\n",
       "      <th>PCP_AL</th>\n",
       "      <th>PCP_CY</th>\n",
       "      <th>PCP_AR</th>\n",
       "      <th>PCP_HB</th>\n",
       "      <th>PCP_HL</th>\n",
       "      <th>...</th>\n",
       "      <th>PCP_SA_BU</th>\n",
       "      <th>PCP_SA_EX</th>\n",
       "      <th>PCP_SA_IN</th>\n",
       "      <th>PCP_TN</th>\n",
       "      <th>PCP_SM</th>\n",
       "      <th>PCP_Z1</th>\n",
       "      <th>PCP_Z2</th>\n",
       "      <th>PCP_Z3</th>\n",
       "      <th>PCP_Z4</th>\n",
       "      <th>PCP_Z5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.093</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.837</td>\n",
       "      <td>0.302</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.186</td>\n",
       "      <td>...</td>\n",
       "      <td>0.558</td>\n",
       "      <td>0.209</td>\n",
       "      <td>0.186</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.721</td>\n",
       "      <td>0.799</td>\n",
       "      <td>-0.798</td>\n",
       "      <td>0.434</td>\n",
       "      <td>-0.325</td>\n",
       "      <td>-0.490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.124</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.794</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.443</td>\n",
       "      <td>0.371</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.526</td>\n",
       "      <td>0.340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.227</td>\n",
       "      <td>0.268</td>\n",
       "      <td>0.691</td>\n",
       "      <td>0.618</td>\n",
       "      <td>-0.324</td>\n",
       "      <td>0.238</td>\n",
       "      <td>-0.341</td>\n",
       "      <td>-0.152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.107</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.179</td>\n",
       "      <td>...</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.357</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-0.263</td>\n",
       "      <td>-0.330</td>\n",
       "      <td>-0.695</td>\n",
       "      <td>-0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.250</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.300</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.425</td>\n",
       "      <td>0.300</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.600</td>\n",
       "      <td>1.129</td>\n",
       "      <td>-0.597</td>\n",
       "      <td>0.222</td>\n",
       "      <td>-0.022</td>\n",
       "      <td>-0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.150</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.300</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.962</td>\n",
       "      <td>-0.002</td>\n",
       "      <td>-0.257</td>\n",
       "      <td>-0.820</td>\n",
       "      <td>0.281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>0.037</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.926</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.704</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.593</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.704</td>\n",
       "      <td>-0.329</td>\n",
       "      <td>-1.297</td>\n",
       "      <td>0.556</td>\n",
       "      <td>-0.407</td>\n",
       "      <td>-0.425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8087</th>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.381</td>\n",
       "      <td>...</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1.373</td>\n",
       "      <td>-0.309</td>\n",
       "      <td>0.556</td>\n",
       "      <td>-0.242</td>\n",
       "      <td>-0.268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8088</th>\n",
       "      <td>0.260</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0.380</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.360</td>\n",
       "      <td>0.320</td>\n",
       "      <td>...</td>\n",
       "      <td>0.380</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.340</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.887</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>0.129</td>\n",
       "      <td>-0.064</td>\n",
       "      <td>-0.445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8089</th>\n",
       "      <td>0.143</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.857</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.607</td>\n",
       "      <td>0.536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.357</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.607</td>\n",
       "      <td>-0.636</td>\n",
       "      <td>-0.045</td>\n",
       "      <td>0.224</td>\n",
       "      <td>0.318</td>\n",
       "      <td>0.674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8090</th>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.143</td>\n",
       "      <td>1.159</td>\n",
       "      <td>0.873</td>\n",
       "      <td>-0.456</td>\n",
       "      <td>-0.630</td>\n",
       "      <td>-0.307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8091 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PCP_PC  PCP_NC  PCP_NE  PCP_PO  PCP_NP  PCP_AL  PCP_CY  PCP_AR  PCP_HB  \\\n",
       "0      0.093   0.070   0.837   0.302   0.442   0.372   0.000   0.140   0.442   \n",
       "1      0.124   0.082   0.794   0.216   0.443   0.371   0.082   0.093   0.526   \n",
       "2      0.107   0.179   0.714   0.214   0.429   0.393   0.000   0.107   0.429   \n",
       "3      0.250   0.075   0.675   0.275   0.350   0.300   0.000   0.050   0.425   \n",
       "4      0.150   0.317   0.533   0.150   0.367   0.283   0.033   0.067   0.367   \n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "8086   0.037   0.037   0.926   0.444   0.481   0.481   0.111   0.074   0.704   \n",
       "8087   0.190   0.095   0.714   0.381   0.238   0.190   0.095   0.048   0.429   \n",
       "8088   0.260   0.080   0.660   0.380   0.240   0.220   0.020   0.140   0.360   \n",
       "8089   0.143   0.000   0.857   0.107   0.750   0.643   0.393   0.143   0.607   \n",
       "8090   0.286   0.286   0.429   0.429   0.000   0.000   0.000   0.286   0.000   \n",
       "\n",
       "      PCP_HL  ...  PCP_SA_BU  PCP_SA_EX  PCP_SA_IN  PCP_TN  PCP_SM  PCP_Z1  \\\n",
       "0      0.186  ...      0.558      0.209      0.186   0.488   0.721   0.799   \n",
       "1      0.340  ...      0.433      0.330      0.227   0.268   0.691   0.618   \n",
       "2      0.179  ...      0.464      0.429      0.107   0.214   0.357   0.070   \n",
       "3      0.300  ...      0.500      0.250      0.175   0.450   0.600   1.129   \n",
       "4      0.200  ...      0.300      0.533      0.150   0.167   0.350   0.962   \n",
       "...      ...  ...        ...        ...        ...     ...     ...     ...   \n",
       "8086   0.148  ...      0.593      0.148      0.333   0.407   0.704  -0.329   \n",
       "8087   0.381  ...      0.286      0.381      0.381   0.381   0.667   1.373   \n",
       "8088   0.320  ...      0.380      0.400      0.220   0.340   0.520   0.887   \n",
       "8089   0.536  ...      0.357      0.429      0.464   0.179   0.607  -0.636   \n",
       "8090   0.286  ...      0.000      0.571      0.429   0.143   0.143   1.159   \n",
       "\n",
       "      PCP_Z2  PCP_Z3  PCP_Z4  PCP_Z5  \n",
       "0     -0.798   0.434  -0.325  -0.490  \n",
       "1     -0.324   0.238  -0.341  -0.152  \n",
       "2     -0.263  -0.330  -0.695  -0.012  \n",
       "3     -0.597   0.222  -0.022  -0.250  \n",
       "4     -0.002  -0.257  -0.820   0.281  \n",
       "...      ...     ...     ...     ...  \n",
       "8086  -1.297   0.556  -0.407  -0.425  \n",
       "8087  -0.309   0.556  -0.242  -0.268  \n",
       "8088  -0.247   0.129  -0.064  -0.445  \n",
       "8089  -0.045   0.224   0.318   0.674  \n",
       "8090   0.873  -0.456  -0.630  -0.307  \n",
       "\n",
       "[8091 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#correlacion\n",
    "corr_matrix = pd.DataFrame(feature_pcp).corr().abs()\n",
    "upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "to_drop = [column for column in upper.columns if any(upper[column] > 0.9)]\n",
    "corr_matrix_pcp = feature_pcp.drop(to_drop, axis=1)\n",
    "corr_matrix_pcp.to_csv('../data/processed/corr_matrix_pcp.csv', index = False)\n",
    "corr_matrix_pcp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Composition enhanced Transition and Distribution: 189 descriptors [4]\n",
    "\n",
    "Composition-Enhanced Transition and Distribution (CTD) es un enfoque utilizado en bioinformática y química computacional para describir secuencias de aminoácidos o nucleótidos. Se basa en tres principios clave:\n",
    "\n",
    "Composición (Composition): Representa la proporción de cada tipo de aminoácido o nucleótido en una secuencia.\n",
    "Transición (Transition): Describe la frecuencia con la que un tipo de aminoácido/nucleótido cambia a otro a lo largo de la secuencia.\n",
    "Distribución (Distribution): Indica la posición relativa de ciertos tipos de aminoácidos/nucleótidos dentro de la secuencia.\n",
    "\n",
    "Cómo funciona CTD en la descripción de secuencias\n",
    "\n",
    "1. Composición (C - Group 1) \n",
    "\n",
    "- Calcula el porcentaje de cada grupo de aminoácidos en la secuencia.\n",
    "- Se agrupan los aminoácidos según ciertas propiedades fisicoquímicas, como hidrofobicidad o carga.\n",
    "\n",
    "2. Transición (T - Group 2)\n",
    "\n",
    "- Representa la frecuencia de cambio entre dos grupos de aminoácidos consecutivos en la secuencia.\n",
    "- Se calcula como el número de transiciones entre los grupos dividido por el total de transiciones posibles.\n",
    "\n",
    "3. Distribución (D - Group 3)\n",
    "\n",
    "- Indica en qué posiciones de la secuencia aparecen los primeros, el 25%, 50%, 75% y 100% de los residuos de un grupo específico.\n",
    "\n",
    "|Attributes|Group1|Group 2|Group 3|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|Hydrophobicity| R,K,E,D,Q,N| G,A,S,T,P,H,Y| C,L,V,I,M,F,W|\n",
    "|Normalized Vander Waals volume| G,A,S,T,P,D| N,V,E,Q,I,L| M,H,K,F,R,Y,W|\n",
    "|Polarity| L,I,F,W,C,M,V,Y| P,A,T,G,S| H,Q,R,K,N,E,D|\n",
    "|Polarizability| G,A,S,D,T| C,P,N,V,E,Q,I,L| K,M,H,F,R,Y,W|\n",
    "|Charge| K,R| A,N,C,Q,G,H,I,L,M,F,P,S,T,W,Y,V| D,E|\n",
    "|Secondary structure| E,A,L,M,Q,K,R,H| V,I,Y,C,W,F,T| G,N,P,S,D|\n",
    "|Solvent accessibility| A,L,F,C,G,I,V,W| R,K,Q,E,N,D| M,S P,T,H,Y|\n",
    "\n",
    "Los atributos de CTD permiten capturar información estructural y funcional de secuencias biológicas, convirtiéndolas en vectores numéricos que pueden ser utilizados para la predicción y clasificación en bioinformática."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CeTD_HB1</th>\n",
       "      <th>CeTD_HB2</th>\n",
       "      <th>CeTD_HB3</th>\n",
       "      <th>CeTD_VW1</th>\n",
       "      <th>CeTD_VW2</th>\n",
       "      <th>CeTD_VW3</th>\n",
       "      <th>CeTD_PO1</th>\n",
       "      <th>CeTD_PO2</th>\n",
       "      <th>CeTD_PO3</th>\n",
       "      <th>CeTD_PZ1</th>\n",
       "      <th>...</th>\n",
       "      <th>CeTD_0_p_SS3</th>\n",
       "      <th>CeTD_25_p_SS3</th>\n",
       "      <th>CeTD_50_p_SS3</th>\n",
       "      <th>CeTD_75_p_SS3</th>\n",
       "      <th>CeTD_100_p_SS3</th>\n",
       "      <th>CeTD_0_p_SA3</th>\n",
       "      <th>CeTD_25_p_SA3</th>\n",
       "      <th>CeTD_50_p_SA3</th>\n",
       "      <th>CeTD_75_p_SA3</th>\n",
       "      <th>CeTD_100_p_SA3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>25.58</td>\n",
       "      <td>48.84</td>\n",
       "      <td>25.58</td>\n",
       "      <td>50.00</td>\n",
       "      <td>23.68</td>\n",
       "      <td>26.32</td>\n",
       "      <td>32.56</td>\n",
       "      <td>41.86</td>\n",
       "      <td>25.58</td>\n",
       "      <td>44.19</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34.02</td>\n",
       "      <td>40.21</td>\n",
       "      <td>25.77</td>\n",
       "      <td>48.89</td>\n",
       "      <td>27.78</td>\n",
       "      <td>23.33</td>\n",
       "      <td>27.84</td>\n",
       "      <td>37.11</td>\n",
       "      <td>35.05</td>\n",
       "      <td>37.11</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42.86</td>\n",
       "      <td>25.00</td>\n",
       "      <td>32.14</td>\n",
       "      <td>22.22</td>\n",
       "      <td>55.56</td>\n",
       "      <td>22.22</td>\n",
       "      <td>39.29</td>\n",
       "      <td>17.86</td>\n",
       "      <td>42.86</td>\n",
       "      <td>21.43</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32.50</td>\n",
       "      <td>37.50</td>\n",
       "      <td>30.00</td>\n",
       "      <td>44.12</td>\n",
       "      <td>20.59</td>\n",
       "      <td>35.29</td>\n",
       "      <td>30.00</td>\n",
       "      <td>32.50</td>\n",
       "      <td>37.50</td>\n",
       "      <td>37.50</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>31.67</td>\n",
       "      <td>43.33</td>\n",
       "      <td>25.00</td>\n",
       "      <td>21.67</td>\n",
       "      <td>21.67</td>\n",
       "      <td>56.67</td>\n",
       "      <td>28.33</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>7.41</td>\n",
       "      <td>48.15</td>\n",
       "      <td>44.44</td>\n",
       "      <td>52.38</td>\n",
       "      <td>33.33</td>\n",
       "      <td>14.29</td>\n",
       "      <td>51.85</td>\n",
       "      <td>40.74</td>\n",
       "      <td>7.41</td>\n",
       "      <td>29.63</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8087</th>\n",
       "      <td>33.33</td>\n",
       "      <td>47.62</td>\n",
       "      <td>19.05</td>\n",
       "      <td>50.00</td>\n",
       "      <td>22.22</td>\n",
       "      <td>27.78</td>\n",
       "      <td>19.05</td>\n",
       "      <td>42.86</td>\n",
       "      <td>38.10</td>\n",
       "      <td>33.33</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8088</th>\n",
       "      <td>40.00</td>\n",
       "      <td>36.00</td>\n",
       "      <td>24.00</td>\n",
       "      <td>35.71</td>\n",
       "      <td>16.67</td>\n",
       "      <td>47.62</td>\n",
       "      <td>36.00</td>\n",
       "      <td>24.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>28.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8089</th>\n",
       "      <td>17.86</td>\n",
       "      <td>60.71</td>\n",
       "      <td>21.43</td>\n",
       "      <td>57.14</td>\n",
       "      <td>14.29</td>\n",
       "      <td>28.57</td>\n",
       "      <td>25.00</td>\n",
       "      <td>57.14</td>\n",
       "      <td>17.86</td>\n",
       "      <td>17.86</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8090</th>\n",
       "      <td>57.14</td>\n",
       "      <td>42.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>14.29</td>\n",
       "      <td>28.57</td>\n",
       "      <td>57.14</td>\n",
       "      <td>28.57</td>\n",
       "      <td>14.29</td>\n",
       "      <td>57.14</td>\n",
       "      <td>14.29</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8091 rows × 189 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CeTD_HB1  CeTD_HB2  CeTD_HB3  CeTD_VW1  CeTD_VW2  CeTD_VW3  CeTD_PO1  \\\n",
       "0        25.58     48.84     25.58     50.00     23.68     26.32     32.56   \n",
       "1        34.02     40.21     25.77     48.89     27.78     23.33     27.84   \n",
       "2        42.86     25.00     32.14     22.22     55.56     22.22     39.29   \n",
       "3        32.50     37.50     30.00     44.12     20.59     35.29     30.00   \n",
       "4        55.00     25.00     20.00     31.67     43.33     25.00     21.67   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "8086      7.41     48.15     44.44     52.38     33.33     14.29     51.85   \n",
       "8087     33.33     47.62     19.05     50.00     22.22     27.78     19.05   \n",
       "8088     40.00     36.00     24.00     35.71     16.67     47.62     36.00   \n",
       "8089     17.86     60.71     21.43     57.14     14.29     28.57     25.00   \n",
       "8090     57.14     42.86      0.00     14.29     28.57     57.14     28.57   \n",
       "\n",
       "      CeTD_PO2  CeTD_PO3  CeTD_PZ1  ...  CeTD_0_p_SS3  CeTD_25_p_SS3  \\\n",
       "0        41.86     25.58     44.19  ...             0              2   \n",
       "1        37.11     35.05     37.11  ...             0              8   \n",
       "2        17.86     42.86     21.43  ...             0              3   \n",
       "3        32.50     37.50     37.50  ...             0              3   \n",
       "4        21.67     56.67     28.33  ...             0              8   \n",
       "...        ...       ...       ...  ...           ...            ...   \n",
       "8086     40.74      7.41     29.63  ...             0              0   \n",
       "8087     42.86     38.10     33.33  ...             0              1   \n",
       "8088     24.00     40.00     28.00  ...             0              5   \n",
       "8089     57.14     17.86     17.86  ...             0              1   \n",
       "8090     14.29     57.14     14.29  ...             0              5   \n",
       "\n",
       "      CeTD_50_p_SS3  CeTD_75_p_SS3  CeTD_100_p_SS3  CeTD_0_p_SA3  \\\n",
       "0                 5              8              11             0   \n",
       "1                16             24              33             0   \n",
       "2                 6              9              12             0   \n",
       "3                 6              9              13             0   \n",
       "4                16             24              33             0   \n",
       "...             ...            ...             ...           ...   \n",
       "8086              1              1               2             0   \n",
       "8087              3              5               7             0   \n",
       "8088             10             15              20             0   \n",
       "8089              2              3               5             0   \n",
       "8090             10             15              20             0   \n",
       "\n",
       "      CeTD_25_p_SA3  CeTD_50_p_SA3  CeTD_75_p_SA3  CeTD_100_p_SA3  \n",
       "0                 2              4              6               8  \n",
       "1                 5             11             16              22  \n",
       "2                 0              1              2               3  \n",
       "3                 1              3              5               7  \n",
       "4                 2              4              6               9  \n",
       "...             ...            ...            ...             ...  \n",
       "8086              2              4              6               9  \n",
       "8087              2              4              6               8  \n",
       "8088              2              5              8              11  \n",
       "8089              3              6              9              13  \n",
       "8090              3              7             11              15  \n",
       "\n",
       "[8091 rows x 189 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Composition enhanced Transition and Distribution: (CTD) \n",
    "\n",
    "def ctd(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'ctd.csv'\n",
    "  df_out = ctd_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_ctd = ctd(cd_hit_path)\n",
    "feature_ctd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se aplicara la transformación de PCA a las caracteristicas de CTD para reducir su dimensionalidad, manteniendo la mayor varianza posible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def aplicar_pca(data, threshold = 0.90):\n",
    "#     print(\"Aplicando operación: PCA\")\n",
    "#     scaler = StandardScaler()\n",
    "#     X_scaled = scaler.fit_transform(data)\n",
    "#     # Aplicar PCA\n",
    "#     pca = PCA().fit(X_scaled)\n",
    "#     # Calcular la varianza explicada acumulada\n",
    "#     explained_variance = np.cumsum(pca.explained_variance_ratio_)\n",
    "#     num_components = np.argmax(explained_variance >= threshold) + 1\n",
    "#     pca = PCA(n_components=num_components)  # Número de componentes principales\n",
    "#     X_pca = pca.fit_transform(X_scaled)\n",
    "#     pca_columns = [f'PC{i+1}' for i in range(X_pca.shape[1])]  # Nombrar los componentes como PC1, PC2, etc.\n",
    "#     df_pca = pd.DataFrame(X_pca, columns=pca_columns)\n",
    "#     return df_pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pca_ctd = aplicar_pca(feature_ctd)\n",
    "# pca_ctd.to_csv('../data/processed/pca_ctd.csv', index = False)\n",
    "# pca_ctd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Repetitive Residue Information (RRI)\n",
    "\n",
    "Repetitive Residue Information (RRI) es una métrica utilizada en bioinformática para analizar la redundancia y patrones de repetición en la secuencia de aminoácidos de péptidos y proteínas. Evalúa cuán frecuente es la aparición de residuos específicos en una secuencia y cómo esta repetición puede influir en la estructura y función de la biomolécula.\n",
    "\n",
    "### Definición de RRI\n",
    "RRI cuantifica la presencia de residuos repetitivos dentro de una secuencia peptídica. Se basa en la frecuencia de aparición de cada aminoácido y su distribución dentro de la secuencia. Esto permite identificar si una proteína o péptido tiene regiones altamente conservadas o patrones estructurales específicos, como repeticiones en hélices alfa o láminas beta.\n",
    "\n",
    "La fórmula general para calcular RRI es:\n",
    "\n",
    "![alt text](image.png)\n",
    "\n",
    "donde:\n",
    "- 𝑓𝑖 es la frecuencia del aminoácido 𝑖 en la secuencia.\n",
    "- 𝑁 es la longitud total de la secuencia.\n",
    "\n",
    "Esta ecuación es similar a la entropía de Shannon, midiendo la diversidad de residuos en la secuencia.\n",
    "\n",
    "### Aplicaciones de RRI en Bioinformática\n",
    "\n",
    "1. Identificación de motivos estructurales y funcionales\n",
    "\n",
    "- Péptidos con residuos repetitivos pueden formar estructuras secundarias estables, como hélices alfa anfipáticas o láminas beta repetitivas.\n",
    "\n",
    "2. Predicción de propiedades fisicoquímicas\n",
    "\n",
    "- Secuencias con alta repetición pueden afectar la hidrofobicidad, la solubilidad y la interacción con membranas celulares.\n",
    "\n",
    "3. Diseño de péptidos terapéuticos\n",
    "\n",
    "- Se utiliza en el diseño de péptidos antimicrobianos y en la optimización de fármacos basados en péptidos, evitando redundancias que puedan generar respuestas inmunogénicas no deseadas.\n",
    "\n",
    "4. Análisis evolutivo\n",
    "\n",
    "- Ayuda a estudiar la conservación de secuencias en proteínas homólogas y en el desarrollo de alineamientos múltiples.\n",
    "\n",
    "El RRI es una métrica valiosa para analizar la redundancia de residuos en secuencias peptídicas. Su aplicación en bioinformática estructural, diseño de péptidos y análisis evolutivo permite mejorar la comprensión de la relación entre la secuencia y la función de las proteínas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RRI_A</th>\n",
       "      <th>RRI_C</th>\n",
       "      <th>RRI_D</th>\n",
       "      <th>RRI_E</th>\n",
       "      <th>RRI_F</th>\n",
       "      <th>RRI_G</th>\n",
       "      <th>RRI_H</th>\n",
       "      <th>RRI_I</th>\n",
       "      <th>RRI_K</th>\n",
       "      <th>RRI_L</th>\n",
       "      <th>RRI_M</th>\n",
       "      <th>RRI_N</th>\n",
       "      <th>RRI_P</th>\n",
       "      <th>RRI_Q</th>\n",
       "      <th>RRI_R</th>\n",
       "      <th>RRI_S</th>\n",
       "      <th>RRI_T</th>\n",
       "      <th>RRI_V</th>\n",
       "      <th>RRI_W</th>\n",
       "      <th>RRI_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.44</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.67</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.40</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.57</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.62</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.40</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8087</th>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8088</th>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.33</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8089</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8090</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8091 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RRI_A  RRI_C  RRI_D  RRI_E  RRI_F  RRI_G  RRI_H  RRI_I  RRI_K  RRI_L  \\\n",
       "0      3.00    1.0   1.00   1.00    1.0   1.29    0.0    0.0   1.00    0.0   \n",
       "1      1.44    1.0   1.25   0.00    1.0   1.50    1.0    1.0   1.00    1.0   \n",
       "2      1.67    1.0   1.00   1.00    0.0   1.00    0.0    1.0   1.00    1.0   \n",
       "3      1.50    1.0   1.00   1.00    1.0   1.00    1.0    1.0   1.40    1.0   \n",
       "4      1.57    0.0   1.00   1.62    1.0   1.00    1.0    1.0   1.40    1.0   \n",
       "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "8086   1.00    1.0   0.00   1.00    0.0   1.00    0.0    1.0   0.00    1.0   \n",
       "8087   1.00    1.0   0.00   1.00    1.0   1.00    1.0    0.0   1.00    0.0   \n",
       "8088   1.00    1.0   1.00   1.00    1.0   1.00    0.0    1.0   1.33    0.0   \n",
       "8089   0.00    0.0   0.00   0.00    1.0   2.50    0.0    1.0   0.00    1.0   \n",
       "8090   0.00    0.0   0.00   2.00    0.0   0.00    0.0    0.0   2.00    0.0   \n",
       "\n",
       "      RRI_M  RRI_N  RRI_P  RRI_Q  RRI_R  RRI_S  RRI_T  RRI_V  RRI_W  RRI_Y  \n",
       "0       0.0   1.00   0.00    0.0    2.0    1.0    1.0    1.0    1.0    1.0  \n",
       "1       0.0   1.15   1.00    1.0    1.0    1.0    1.0    1.0    0.0    1.0  \n",
       "2       0.0   1.00   0.00    1.0    0.0    1.0    0.0    1.0    1.0    1.0  \n",
       "3       0.0   1.00   0.00    0.0    1.0    1.0    1.0    1.0    0.0    0.0  \n",
       "4       1.0   1.00   1.00    1.0    1.0    1.0    1.0    1.0    0.0    1.0  \n",
       "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  \n",
       "8086    0.0   0.00   1.00    0.0    1.0    1.0    1.0    1.0    0.0    1.0  \n",
       "8087    0.0   1.00   1.00    0.0    1.0    1.0    1.0    0.0    0.0    0.0  \n",
       "8088    0.0   1.00   1.00    1.0    1.0    1.0    1.0    1.0    0.0    1.0  \n",
       "8089    0.0   0.00   1.18    1.0    1.0    1.0    0.0    1.0    0.0    1.0  \n",
       "8090    0.0   0.00   0.00    0.0    0.0    1.0    0.0    0.0    0.0    1.0  \n",
       "\n",
       "[8091 rows x 20 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def rri(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'rri.csv'\n",
    "  df_out = rri_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_rri = rri(cd_hit_path)\n",
    "feature_rri.to_csv('../data/processed/feature_rri.csv', index = False)\n",
    "feature_rri"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analisis ce Componentes Principales para las Pfeatures\n",
    "En este apartado se fusionaran los 4 dataframes con las pfeatures para hacer un análisis de correlacion y PCA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8091, 239)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Se concatenan los 4 dataframes con las pfeatures:\n",
    "df_pfeatures = feature_pcp.merge(feature_ctd, how='inner', left_index=True, right_index=True)\n",
    "df_pfeatures = df_pfeatures.merge(feature_rri, how='inner', left_index=True, right_index=True)\n",
    "df_pfeatures.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlacion: Descarta características redundantes con alta correlación entre sí."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8091, 128)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#correlacion\n",
    "corr_matrix = pd.DataFrame(df_pfeatures).corr().abs()\n",
    "upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "to_drop = [column for column in upper.columns if any(upper[column] > 0.9)]\n",
    "corr_matrix_pfeatures = df_pfeatures.drop(to_drop, axis=1)\n",
    "corr_matrix_pfeatures.to_csv('../data/processed/corr_matrix_pfeatures.csv', index = False)\n",
    "corr_matrix_pfeatures.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se aplicara la transformación de PCA a las pFeatures para reducir su dimensionalidad, manteniendo la mayor varianza posible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Curva de porcentaje de varianza acumulada\n",
    "# total_components = df_pfeatures.shape[1]\n",
    "\n",
    "# pca = PCA()\n",
    "# pca_fit = pca.fit_transform(df_pfeatures)\n",
    "\n",
    "# plt.plot(np.cumsum(pca.explained_variance_ratio_)*100)\n",
    "# plt.title('PCA Analysis')\n",
    "# plt.xlabel('N-th Principal Component')\n",
    "# plt.ylabel('% Cumulative Variance Explained')\n",
    "# plt.xticks(np.arange(0,total_components,20))\n",
    "\n",
    "# labels = np.cumsum(pca.explained_variance_ratio_)*100\n",
    "# for i in range(min(6, total_components)):\n",
    "#   plt.text(i,labels[i],str(format(labels[i],'.0f'))+'%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Se muestra la composición de los 20 componentes principales por cada una de las pFeatures\n",
    "# num_components = 20\n",
    "# pc_df = pd.DataFrame(abs(pca.components_[:num_components]), columns = df_pfeatures.columns)\n",
    "\n",
    "# pc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pca_pfeatures = aplicar_pca(df_pfeatures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pca_pfeatures.index = seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCP_PC</th>\n",
       "      <th>PCP_NC</th>\n",
       "      <th>PCP_NE</th>\n",
       "      <th>PCP_PO</th>\n",
       "      <th>PCP_NP</th>\n",
       "      <th>PCP_AL</th>\n",
       "      <th>PCP_CY</th>\n",
       "      <th>PCP_AR</th>\n",
       "      <th>PCP_AC</th>\n",
       "      <th>PCP_BS</th>\n",
       "      <th>...</th>\n",
       "      <th>RRI_M</th>\n",
       "      <th>RRI_N</th>\n",
       "      <th>RRI_P</th>\n",
       "      <th>RRI_Q</th>\n",
       "      <th>RRI_R</th>\n",
       "      <th>RRI_S</th>\n",
       "      <th>RRI_T</th>\n",
       "      <th>RRI_V</th>\n",
       "      <th>RRI_W</th>\n",
       "      <th>RRI_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.093</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.837</td>\n",
       "      <td>0.302</td>\n",
       "      <td>0.442</td>\n",
       "      <td>0.372</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.070</td>\n",
       "      <td>0.093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.124</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.794</td>\n",
       "      <td>0.216</td>\n",
       "      <td>0.443</td>\n",
       "      <td>0.371</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.124</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.107</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.179</td>\n",
       "      <td>0.107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.250</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.275</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.300</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.150</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.317</td>\n",
       "      <td>0.150</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8086</th>\n",
       "      <td>0.037</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.926</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.037</td>\n",
       "      <td>0.037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8087</th>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.381</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.190</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.095</td>\n",
       "      <td>0.190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8088</th>\n",
       "      <td>0.260</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.660</td>\n",
       "      <td>0.380</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.140</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8089</th>\n",
       "      <td>0.143</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.857</td>\n",
       "      <td>0.107</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.143</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.18</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8090</th>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8091 rows × 239 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PCP_PC  PCP_NC  PCP_NE  PCP_PO  PCP_NP  PCP_AL  PCP_CY  PCP_AR  PCP_AC  \\\n",
       "0      0.093   0.070   0.837   0.302   0.442   0.372   0.000   0.140   0.070   \n",
       "1      0.124   0.082   0.794   0.216   0.443   0.371   0.082   0.093   0.082   \n",
       "2      0.107   0.179   0.714   0.214   0.429   0.393   0.000   0.107   0.179   \n",
       "3      0.250   0.075   0.675   0.275   0.350   0.300   0.000   0.050   0.075   \n",
       "4      0.150   0.317   0.533   0.150   0.367   0.283   0.033   0.067   0.317   \n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "8086   0.037   0.037   0.926   0.444   0.481   0.481   0.111   0.074   0.037   \n",
       "8087   0.190   0.095   0.714   0.381   0.238   0.190   0.095   0.048   0.095   \n",
       "8088   0.260   0.080   0.660   0.380   0.240   0.220   0.020   0.140   0.080   \n",
       "8089   0.143   0.000   0.857   0.107   0.750   0.643   0.393   0.143   0.000   \n",
       "8090   0.286   0.286   0.429   0.429   0.000   0.000   0.000   0.286   0.286   \n",
       "\n",
       "      PCP_BS  ...  RRI_M  RRI_N  RRI_P  RRI_Q  RRI_R  RRI_S  RRI_T  RRI_V  \\\n",
       "0      0.093  ...    0.0   1.00   0.00    0.0    2.0    1.0    1.0    1.0   \n",
       "1      0.124  ...    0.0   1.15   1.00    1.0    1.0    1.0    1.0    1.0   \n",
       "2      0.107  ...    0.0   1.00   0.00    1.0    0.0    1.0    0.0    1.0   \n",
       "3      0.250  ...    0.0   1.00   0.00    0.0    1.0    1.0    1.0    1.0   \n",
       "4      0.150  ...    1.0   1.00   1.00    1.0    1.0    1.0    1.0    1.0   \n",
       "...      ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "8086   0.037  ...    0.0   0.00   1.00    0.0    1.0    1.0    1.0    1.0   \n",
       "8087   0.190  ...    0.0   1.00   1.00    0.0    1.0    1.0    1.0    0.0   \n",
       "8088   0.260  ...    0.0   1.00   1.00    1.0    1.0    1.0    1.0    1.0   \n",
       "8089   0.143  ...    0.0   0.00   1.18    1.0    1.0    1.0    0.0    1.0   \n",
       "8090   0.286  ...    0.0   0.00   0.00    0.0    0.0    1.0    0.0    0.0   \n",
       "\n",
       "      RRI_W  RRI_Y  \n",
       "0       1.0    1.0  \n",
       "1       0.0    1.0  \n",
       "2       1.0    1.0  \n",
       "3       0.0    0.0  \n",
       "4       0.0    1.0  \n",
       "...     ...    ...  \n",
       "8086    0.0    1.0  \n",
       "8087    0.0    0.0  \n",
       "8088    0.0    1.0  \n",
       "8089    0.0    1.0  \n",
       "8090    0.0    1.0  \n",
       "\n",
       "[8091 rows x 239 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pfeatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Se guarda el DF\n",
    "\n",
    "df_pfeatures.to_csv('../data/processed/DTCPep_pfeatures.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusiones\n",
    "En la primera parte de esta actividad hicimos una ingeniería de características a las secuencias de aminoácidos de los péptidos, para depurar aquellos péptidos que contenían caracteres que podrían generar un sesgo al momento de que se haga el modelo predictivo. \n",
    "\n",
    "Posteriormente se utilizó la librería Pfeatures sobre las secuencias de aminoácidos ya depuradas, para obtener diferentes características fisicoquímicas. De estas a su vez se obtuvieron los 20 componentes principales que dan un valor arriba del 90% en la explicación de la varianza de las características.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Referencias\n",
    "\n",
    "1. INTERNATIONAL UNION OF PURE AND APPLIED CHEMISTRY (IUPAC). https://iupac.qmul.ac.uk/AminoAcid/A2021.html#AA20\n",
    "2. Fu, L., Niu, B., Zhu, Z., Wu, S., & Li, W. (2012). CD-HIT: accelerated for clustering the next-generation sequencing data. Bioinformatics (Oxford, England), 28(23), 3150–3152. https://doi.org/10.1093/bioinformatics/bts565\n",
    "3. Raghava, S. P. a. D. a. P. D. K. N. S. G. M. R. K. P. a. V. P. a. L. H. K. S. J. P. G. (n.d.). Pfeature: A web server for computing protein and peptide features. https://webs.iiitd.edu.in/raghava/pfeature/\n",
    "4. Dubchak I, et al. Proceedings of the National Academy of Sciences of the United States of America"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
