{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Instituto Tecnológico y de Estudios Superiores de Monterrey\n",
    "## Maestría en Inteligencia Artificial Aplicada\n",
    "### Proyecto Integrador (Gpo 10) - TC5035.10\n",
    "\n",
    "### **Proyecto: Diseño Acelerado de Fármacos**\n",
    "\n",
    "### Avance 2: Ingeniería de características\n",
    "\n",
    "#### **Docentes:**\n",
    "- Dra. Grettel Barceló Alonso - Profesor Titular\n",
    "- Dr. Luis Eduardo Falcón Morales - Profesor Titular\n",
    "- Dr. Horacio Martinez Alfaro  – Profesor Tutor\n",
    "\n",
    "#### **Miembros del equipo:**\n",
    " - Jose Luis Artigas Arroyo - A01794906\n",
    " - José Luis Fabela Portillo - A00748551\n",
    " - Alejandro Emmanuel Silva Ortega - A01794545"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importar librerias\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "import re\n",
    "from pfeature import pcp_wp\n",
    "from pfeature import ctd_wp\n",
    "from pfeature import sep_wp\n",
    "from pfeature import rri_wp\n",
    "from Bio import SeqIO\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sequence</th>\n",
       "      <td>RGLRRLGRKIAHGVKKYGPTVLRIIRIA</td>\n",
       "      <td>KWKLFKKIGIGKFLHSAKKF</td>\n",
       "      <td>KWKLFKKIGPGKFLHSAKKF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence_Length</th>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Origin</th>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Type</th>\n",
       "      <td>Native peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hemolytic_Activity</th>\n",
       "      <td>Human erythrocytes (4% Hemolysis=4 µM; 19% Hem...</td>\n",
       "      <td>Human erythrocytes (Hemolysis 0% =100 μM)</td>\n",
       "      <td>Human erythrocytes: Hemolysis 0% (100 μM)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cytotoxicity</th>\n",
       "      <td>NIH 3T3: IC50=16 µM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Binding_Target</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Affinity</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mechanism</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nature</th>\n",
       "      <td>Anticancer; Antibacterial; Antifungal</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Classification</th>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Helicity</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear_Cyclic</th>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Disulfide_Bond</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N-terminal_Modification</th>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-terminal_Modification</th>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other_Modification</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chiral</th>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mass</th>\n",
       "      <td>368073</td>\n",
       "      <td>274410</td>\n",
       "      <td>272805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Formula</th>\n",
       "      <td>C144H257N51O31</td>\n",
       "      <td>C120H190N30O22</td>\n",
       "      <td>C119H186N30O22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Absent_amino_acids</th>\n",
       "      <td>CDEFMNQSW</td>\n",
       "      <td>CDEMNPQRTVY</td>\n",
       "      <td>CDEMNQRTVY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Common_amino_acids</th>\n",
       "      <td>R</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pI</th>\n",
       "      <td>12.81</td>\n",
       "      <td>11.58</td>\n",
       "      <td>11.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Basic_residues</th>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acidic_residues</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Net_charge</th>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Polar_residues</th>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hydrophobic_residues</th>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hydrophobicity</th>\n",
       "      <td>-20.36</td>\n",
       "      <td>-31.0</td>\n",
       "      <td>-61.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Boman_Index</th>\n",
       "      <td>-6364</td>\n",
       "      <td>-1227</td>\n",
       "      <td>-1719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Half_Life</th>\n",
       "      <td>Mammalian: 1 hour##Yeast: 2 min##E.coli: 2 min</td>\n",
       "      <td>Mammalian: 1.3 hour##Yeast: 3 min##E.coli: 2 min</td>\n",
       "      <td>Mammalian: 1.3 hour##Yeast: 3 min##E.coli: 2 min</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aliphatic_Index</th>\n",
       "      <td>125.36</td>\n",
       "      <td>83.0</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Extinction_Coefficient_cystines</th>\n",
       "      <td>1490</td>\n",
       "      <td>5500</td>\n",
       "      <td>5500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Absorbance_280nm</th>\n",
       "      <td>55.19</td>\n",
       "      <td>289.47</td>\n",
       "      <td>289.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other_Information</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                 0  \\\n",
       "Sequence                                              RGLRRLGRKIAHGVKKYGPTVLRIIRIA   \n",
       "Sequence_Length                                                                 28   \n",
       "Origin                                                                  Ovis aries   \n",
       "Type                                                                Native peptide   \n",
       "Hemolytic_Activity               Human erythrocytes (4% Hemolysis=4 µM; 19% Hem...   \n",
       "Cytotoxicity                                                   NIH 3T3: IC50=16 µM   \n",
       "Binding_Target                                                                 NaN   \n",
       "Affinity                                                                       NaN   \n",
       "Mechanism                                                                      NaN   \n",
       "Nature                                       Anticancer; Antibacterial; Antifungal   \n",
       "Classification                             Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                                       NaN   \n",
       "Linear_Cyclic                                                               Linear   \n",
       "Disulfide_Bond                                                                 NaN   \n",
       "N-terminal_Modification                                                       Free   \n",
       "C-terminal_Modification                                                  Amidation   \n",
       "Other_Modification                                                             NaN   \n",
       "Chiral                                                                           L   \n",
       "Mass                                                                        368073   \n",
       "Formula                                                             C144H257N51O31   \n",
       "Absent_amino_acids                                                       CDEFMNQSW   \n",
       "Common_amino_acids                                                               R   \n",
       "pI                                                                           12.81   \n",
       "Basic_residues                                                                  10   \n",
       "Acidic_residues                                                                  0   \n",
       "Net_charge                                                                      10   \n",
       "Polar_residues                                                                   6   \n",
       "Hydrophobic_residues                                                            11   \n",
       "Hydrophobicity                                                              -20.36   \n",
       "Boman_Index                                                                  -6364   \n",
       "Half_Life                           Mammalian: 1 hour##Yeast: 2 min##E.coli: 2 min   \n",
       "Aliphatic_Index                                                             125.36   \n",
       "Extinction_Coefficient_cystines                                               1490   \n",
       "Absorbance_280nm                                                             55.19   \n",
       "Other_Information                                                              NaN   \n",
       "VEGF                                                                             0   \n",
       "\n",
       "                                                                                1  \\\n",
       "Sequence                                                     KWKLFKKIGIGKFLHSAKKF   \n",
       "Sequence_Length                                                                20   \n",
       "Origin                                           Ceropin-Magainin hybrid peptides   \n",
       "Type                                                            Synthetic peptide   \n",
       "Hemolytic_Activity                      Human erythrocytes (Hemolysis 0% =100 μM)   \n",
       "Cytotoxicity                                      NIH-3T3 fibroblast: IC50>100 μM   \n",
       "Binding_Target                                                                NaN   \n",
       "Affinity                                                                      NaN   \n",
       "Mechanism                                                                     NaN   \n",
       "Nature                                                  Anticancer; Antibacterial   \n",
       "Classification                            Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                                      NaN   \n",
       "Linear_Cyclic                                                              Linear   \n",
       "Disulfide_Bond                                                                NaN   \n",
       "N-terminal_Modification                                                      Free   \n",
       "C-terminal_Modification                                                 Amidation   \n",
       "Other_Modification                                                            NaN   \n",
       "Chiral                                                                          L   \n",
       "Mass                                                                       274410   \n",
       "Formula                                                            C120H190N30O22   \n",
       "Absent_amino_acids                                                    CDEMNPQRTVY   \n",
       "Common_amino_acids                                                              K   \n",
       "pI                                                                          11.58   \n",
       "Basic_residues                                                                  8   \n",
       "Acidic_residues                                                                 0   \n",
       "Net_charge                                                                      8   \n",
       "Polar_residues                                                                  3   \n",
       "Hydrophobic_residues                                                            9   \n",
       "Hydrophobicity                                                              -31.0   \n",
       "Boman_Index                                                                 -1227   \n",
       "Half_Life                        Mammalian: 1.3 hour##Yeast: 3 min##E.coli: 2 min   \n",
       "Aliphatic_Index                                                              83.0   \n",
       "Extinction_Coefficient_cystines                                              5500   \n",
       "Absorbance_280nm                                                           289.47   \n",
       "Other_Information                                                             NaN   \n",
       "VEGF                                                                            0   \n",
       "\n",
       "                                                                                2  \n",
       "Sequence                                                     KWKLFKKIGPGKFLHSAKKF  \n",
       "Sequence_Length                                                                20  \n",
       "Origin                                           Ceropin-Magainin hybrid peptides  \n",
       "Type                                                            Synthetic peptide  \n",
       "Hemolytic_Activity                      Human erythrocytes: Hemolysis 0% (100 μM)  \n",
       "Cytotoxicity                                      NIH-3T3 fibroblast: IC50>100 μM  \n",
       "Binding_Target                                                                NaN  \n",
       "Affinity                                                                      NaN  \n",
       "Mechanism                                                                     NaN  \n",
       "Nature                                                  Anticancer; Antibacterial  \n",
       "Classification                            Tumor active peptide##Membrane-targeted  \n",
       "Helicity                                                                      NaN  \n",
       "Linear_Cyclic                                                              Linear  \n",
       "Disulfide_Bond                                                                NaN  \n",
       "N-terminal_Modification                                                      Free  \n",
       "C-terminal_Modification                                                 Amidation  \n",
       "Other_Modification                                                            NaN  \n",
       "Chiral                                                                          L  \n",
       "Mass                                                                       272805  \n",
       "Formula                                                            C119H186N30O22  \n",
       "Absent_amino_acids                                                     CDEMNQRTVY  \n",
       "Common_amino_acids                                                              K  \n",
       "pI                                                                          11.58  \n",
       "Basic_residues                                                                  8  \n",
       "Acidic_residues                                                                 0  \n",
       "Net_charge                                                                      8  \n",
       "Polar_residues                                                                  3  \n",
       "Hydrophobic_residues                                                            8  \n",
       "Hydrophobicity                                                              -61.5  \n",
       "Boman_Index                                                                 -1719  \n",
       "Half_Life                        Mammalian: 1.3 hour##Yeast: 3 min##E.coli: 2 min  \n",
       "Aliphatic_Index                                                              63.5  \n",
       "Extinction_Coefficient_cystines                                              5500  \n",
       "Absorbance_280nm                                                           289.47  \n",
       "Other_Information                                                             NaN  \n",
       "VEGF                                                                            0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#leer csv con dataset preprocesado\n",
    "df_peptidos = pd.read_csv('../data/processed/peptide_processed.csv')\n",
    "df_peptidos.head(3).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eliminacion de columnas redundantes\n",
    "Despues de un analis mas profundo, y una revision con el experto en el tema, se procedera a la eliminacion de columnas redundantes, ya que estos valores pueden ser obtenidos direcvtamente de la secuencia de peptidos, estas columnas son\n",
    "\n",
    "- Chiral: Los aminoácidos L/D consisten en péptidos.\n",
    "- Mass: Masa\n",
    "- Formula: Formula del péptido\n",
    "- Absent_amino_acids: Aminoacidos no incluidos\n",
    "- Common_amino_acids: Animoacido mas comun en la sequencia\n",
    "- pI: Punto isoeléctrico\n",
    "- Basic_residues: Número de residuos base\n",
    "- Acidic_residues: Número de residuos ácidos\n",
    "- Net_charge: Carga Neta\n",
    "- Polar_residues: Número de residuos polares\n",
    "- Hydrophobic_residues: Número de residuos hidrofóbicos\n",
    "- Hydrophobicity: Hidrofobicidad\n",
    "- Boman_Index: Índice de Boman\n",
    "- Half_Life: Vida media\n",
    "- Aliphatic_Index: Índice alifático\n",
    "- Extinction_Coefficient_cystines: Coeficiente de extinción de cistinas\n",
    "- Absorbance_280nm: Absorbencia a 280nm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_cols = ['Chiral','Mass','Formula','Absent_amino_acids','Common_amino_acids','pI','Basic_residues','Acidic_residues','Net_charge','Polar_residues','Hydrophobic_residues','Hydrophobicity','Boman_Index',\n",
    "             'Half_Life','Aliphatic_Index','Extinction_Coefficient_cystines','Absorbance_280nm']\n",
    "df_peptidos.drop(drop_cols, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sequence</th>\n",
       "      <td>RGLRRLGRKIAHGVKKYGPTVLRIIRIA</td>\n",
       "      <td>KWKLFKKIGIGKFLHSAKKF</td>\n",
       "      <td>KWKLFKKIGPGKFLHSAKKF</td>\n",
       "      <td>KWKLFKKIKFLHSAKKF</td>\n",
       "      <td>KWKLFKKIPKFLHSAKKF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence_Length</th>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Origin</th>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "      <td>Ceropin-Magainin hybrid peptides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Type</th>\n",
       "      <td>Native peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "      <td>Synthetic peptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hemolytic_Activity</th>\n",
       "      <td>Human erythrocytes (4% Hemolysis=4 µM; 19% Hem...</td>\n",
       "      <td>Human erythrocytes (Hemolysis 0% =100 μM)</td>\n",
       "      <td>Human erythrocytes: Hemolysis 0% (100 μM)</td>\n",
       "      <td>Human erythrocytes: Hemolysis 0% (100 μM)</td>\n",
       "      <td>Human erythrocytes: Hemolysis 0% (100 μM)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cytotoxicity</th>\n",
       "      <td>NIH 3T3: IC50=16 µM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "      <td>NIH-3T3 fibroblast: IC50&gt;100 μM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Binding_Target</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Affinity</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mechanism</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nature</th>\n",
       "      <td>Anticancer; Antibacterial; Antifungal</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "      <td>Anticancer; Antibacterial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Classification</th>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "      <td>Tumor active peptide##Membrane-targeted</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Helicity</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear_Cyclic</th>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "      <td>Linear</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Disulfide_Bond</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N-terminal_Modification</th>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "      <td>Free</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-terminal_Modification</th>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "      <td>Amidation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other_Modification</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other_Information</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         0  \\\n",
       "Sequence                                      RGLRRLGRKIAHGVKKYGPTVLRIIRIA   \n",
       "Sequence_Length                                                         28   \n",
       "Origin                                                          Ovis aries   \n",
       "Type                                                        Native peptide   \n",
       "Hemolytic_Activity       Human erythrocytes (4% Hemolysis=4 µM; 19% Hem...   \n",
       "Cytotoxicity                                           NIH 3T3: IC50=16 µM   \n",
       "Binding_Target                                                         NaN   \n",
       "Affinity                                                               NaN   \n",
       "Mechanism                                                              NaN   \n",
       "Nature                               Anticancer; Antibacterial; Antifungal   \n",
       "Classification                     Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                               NaN   \n",
       "Linear_Cyclic                                                       Linear   \n",
       "Disulfide_Bond                                                         NaN   \n",
       "N-terminal_Modification                                               Free   \n",
       "C-terminal_Modification                                          Amidation   \n",
       "Other_Modification                                                     NaN   \n",
       "Other_Information                                                      NaN   \n",
       "VEGF                                                                     0   \n",
       "\n",
       "                                                                 1  \\\n",
       "Sequence                                      KWKLFKKIGIGKFLHSAKKF   \n",
       "Sequence_Length                                                 20   \n",
       "Origin                            Ceropin-Magainin hybrid peptides   \n",
       "Type                                             Synthetic peptide   \n",
       "Hemolytic_Activity       Human erythrocytes (Hemolysis 0% =100 μM)   \n",
       "Cytotoxicity                       NIH-3T3 fibroblast: IC50>100 μM   \n",
       "Binding_Target                                                 NaN   \n",
       "Affinity                                                       NaN   \n",
       "Mechanism                                                      NaN   \n",
       "Nature                                   Anticancer; Antibacterial   \n",
       "Classification             Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                       NaN   \n",
       "Linear_Cyclic                                               Linear   \n",
       "Disulfide_Bond                                                 NaN   \n",
       "N-terminal_Modification                                       Free   \n",
       "C-terminal_Modification                                  Amidation   \n",
       "Other_Modification                                             NaN   \n",
       "Other_Information                                              NaN   \n",
       "VEGF                                                             0   \n",
       "\n",
       "                                                                 2  \\\n",
       "Sequence                                      KWKLFKKIGPGKFLHSAKKF   \n",
       "Sequence_Length                                                 20   \n",
       "Origin                            Ceropin-Magainin hybrid peptides   \n",
       "Type                                             Synthetic peptide   \n",
       "Hemolytic_Activity       Human erythrocytes: Hemolysis 0% (100 μM)   \n",
       "Cytotoxicity                       NIH-3T3 fibroblast: IC50>100 μM   \n",
       "Binding_Target                                                 NaN   \n",
       "Affinity                                                       NaN   \n",
       "Mechanism                                                      NaN   \n",
       "Nature                                   Anticancer; Antibacterial   \n",
       "Classification             Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                       NaN   \n",
       "Linear_Cyclic                                               Linear   \n",
       "Disulfide_Bond                                                 NaN   \n",
       "N-terminal_Modification                                       Free   \n",
       "C-terminal_Modification                                  Amidation   \n",
       "Other_Modification                                             NaN   \n",
       "Other_Information                                              NaN   \n",
       "VEGF                                                             0   \n",
       "\n",
       "                                                                 3  \\\n",
       "Sequence                                         KWKLFKKIKFLHSAKKF   \n",
       "Sequence_Length                                                 17   \n",
       "Origin                            Ceropin-Magainin hybrid peptides   \n",
       "Type                                             Synthetic peptide   \n",
       "Hemolytic_Activity       Human erythrocytes: Hemolysis 0% (100 μM)   \n",
       "Cytotoxicity                       NIH-3T3 fibroblast: IC50>100 μM   \n",
       "Binding_Target                                                 NaN   \n",
       "Affinity                                                       NaN   \n",
       "Mechanism                                                      NaN   \n",
       "Nature                                   Anticancer; Antibacterial   \n",
       "Classification             Tumor active peptide##Membrane-targeted   \n",
       "Helicity                                                       NaN   \n",
       "Linear_Cyclic                                               Linear   \n",
       "Disulfide_Bond                                                 NaN   \n",
       "N-terminal_Modification                                       Free   \n",
       "C-terminal_Modification                                  Amidation   \n",
       "Other_Modification                                             NaN   \n",
       "Other_Information                                              NaN   \n",
       "VEGF                                                             0   \n",
       "\n",
       "                                                                 4  \n",
       "Sequence                                        KWKLFKKIPKFLHSAKKF  \n",
       "Sequence_Length                                                 18  \n",
       "Origin                            Ceropin-Magainin hybrid peptides  \n",
       "Type                                             Synthetic peptide  \n",
       "Hemolytic_Activity       Human erythrocytes: Hemolysis 0% (100 μM)  \n",
       "Cytotoxicity                       NIH-3T3 fibroblast: IC50>100 μM  \n",
       "Binding_Target                                                 NaN  \n",
       "Affinity                                                       NaN  \n",
       "Mechanism                                                      NaN  \n",
       "Nature                                   Anticancer; Antibacterial  \n",
       "Classification             Tumor active peptide##Membrane-targeted  \n",
       "Helicity                                                       NaN  \n",
       "Linear_Cyclic                                               Linear  \n",
       "Disulfide_Bond                                                 NaN  \n",
       "N-terminal_Modification                                       Free  \n",
       "C-terminal_Modification                                  Amidation  \n",
       "Other_Modification                                             NaN  \n",
       "Other_Information                                              NaN  \n",
       "VEGF                                                             0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos.head(5).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analisis de secuencias de péptidos\n",
    "\n",
    "De acuerdo a la IUPAC[1], los aminoacidos se clasifican de acuerdo a la siguiente tabla\n",
    "\n",
    "\n",
    "|symbol|Amino acid|\n",
    "| :-:|:-:|\n",
    "|A\t| alanine|\n",
    "|B\t| aspartic acid or asparagine|\n",
    "|C\t| cysteine|\n",
    "|D\t| aspartic acid|\n",
    "|E\t| glutamic acid|\n",
    "|F\t| phenylalanine|\n",
    "|G\t| glycine|\n",
    "|H\t| histidine|\n",
    "|I\t| isoleucine|\n",
    "|K\t| lysine|\n",
    "|L\t| leucine|\n",
    "|M\t| methionine|\n",
    "|N\t| asparagine|\n",
    "|P\t| proline|\n",
    "|Q\t| glutamine|\n",
    "|R\t| arginine|\n",
    "|S\t| serine|\n",
    "|T\t| threonine|\n",
    "|U  | selenocysteine|\n",
    "|V\t| valine|\n",
    "|W\t| tryptophan|\n",
    "|X  | Desconocido u 'otro'|\n",
    "|Y\t| tyrosine|\n",
    "|Z\t| glutamic acid or glutamine|\n",
    "\n",
    "Por lo cual, se eliminaran de la secuencia los caractares diferentes a los indicados en esta lista, de igual forma se actualizara la longitud de la secuencia para hacerla coincidir con la longitud despues de eliminar los caracteres invalidos, se eliminaran los registros con cadenas que contengan \"-\" y numeros (1-9) ya que estas representan peptidos modificados y contienen compuestos quimicos los cuales modifican el péptido y estos no seran considerados para el modelo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                RGLRRLGRKIAHGVKKYGPTVLRIIRIA\n",
       "1                        KWKLFKKIGIGKFLHSAKKF\n",
       "2                        KWKLFKKIGPGKFLHSAKKF\n",
       "3                           KWKLFKKIKFLHSAKKF\n",
       "4                          KWKLFKKIPKFLHSAKKF\n",
       "                        ...                  \n",
       "5846                 FLPLLAGLAANFLPKIICKIARKC\n",
       "5847                  RIIDRLWLVRRPOKPKFVLVWVL\n",
       "5848                               FHAVPQSFYT\n",
       "5849                            KLWCKSSQVPQSR\n",
       "5850    TKEQKEQIAKATGLTTKQVRNWYVQLNASIKVCMCSC\n",
       "Name: Sequence, Length: 5851, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos[\"Sequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                RGLRRLGRKIAHGVKKYGPTVLRIIRIA\n",
       "1                        KWKLFKKIGIGKFLHSAKKF\n",
       "2                        KWKLFKKIGPGKFLHSAKKF\n",
       "3                           KWKLFKKIKFLHSAKKF\n",
       "4                          KWKLFKKIPKFLHSAKKF\n",
       "                        ...                  \n",
       "5846                 FLPLLAGLAANFLPKIICKIARKC\n",
       "5847                  RIIDRLWLVRRPOKPKFVLVWVL\n",
       "5848                               FHAVPQSFYT\n",
       "5849                            KLWCKSSQVPQSR\n",
       "5850    TKEQKEQIAKATGLTTKQVRNWYVQLNASIKVCMCSC\n",
       "Name: Sequence, Length: 5851, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peptidos[\"Sequence\"].map(lambda x: re.sub(r\"[\\\"]\",'',x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>V</td>\n",
       "      <td>P</td>\n",
       "      <td>R</td>\n",
       "      <td>K</td>\n",
       "      <td>H</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>Φ</td>\n",
       "      <td>(</td>\n",
       "      <td>)</td>\n",
       "      <td>]</td>\n",
       "      <td>[</td>\n",
       "      <td>b</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>3014</td>\n",
       "      <td>2964</td>\n",
       "      <td>3081</td>\n",
       "      <td>3858</td>\n",
       "      <td>2932</td>\n",
       "      <td>2435</td>\n",
       "      <td>3207</td>\n",
       "      <td>3390</td>\n",
       "      <td>1033</td>\n",
       "      <td>1684</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>51.51</td>\n",
       "      <td>50.66</td>\n",
       "      <td>52.66</td>\n",
       "      <td>65.94</td>\n",
       "      <td>50.11</td>\n",
       "      <td>41.62</td>\n",
       "      <td>54.81</td>\n",
       "      <td>57.94</td>\n",
       "      <td>17.66</td>\n",
       "      <td>28.78</td>\n",
       "      <td>...</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0      1      2      3      4      5      6      7      8   \\\n",
       "Character      A      G      I      L      V      P      R      K      H   \n",
       "Count       3014   2964   3081   3858   2932   2435   3207   3390   1033   \n",
       "%          51.51  50.66  52.66  65.94  50.11  41.62  54.81  57.94  17.66   \n",
       "\n",
       "              9   ...    54    55    56    57    58    59    60    61    62  \\\n",
       "Character      T  ...     Φ     (     )     ]     [     b     6     2     1   \n",
       "Count       1684  ...    11     4     4     1     1     1     1     1     1   \n",
       "%          28.78  ...  0.19  0.07  0.07  0.02  0.02  0.02  0.02  0.02  0.02   \n",
       "\n",
       "             63  \n",
       "Character     7  \n",
       "Count         1  \n",
       "%          0.02  \n",
       "\n",
       "[3 rows x 64 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Crear un diccionario para contar la aparición de cada carácter de la secuencia de aminoácidos\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in df_peptidos['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / df_peptidos.shape[0]).round(4) *100\n",
    "\n",
    "# Número de péptidos por caracter\n",
    "char_count_df.T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Es importante identificar los caracteres contenidos en la secuencia de aminoácidos de los péptidos, por tal motivo vamos a analizar su ocurrencia en el total de péptidos. La lista utilizada es la que se describio anteriormente (IUPAC), sin embargo, para obtener las características con la librería de Pfeatures no se reconocen 4 de estos: B, U, X, Z. Asímismo existen otros caracteres no incluidos en IUPAC por lo que es necesario analizar que tan frecuentes son."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de aminoácidos IUPAC: 24\n",
      "Total de aminoácidos Pfeatures: 20\n",
      "Total de aminoácidos que están en UIPAC pero no reconoce Pfeatures: 4\n"
     ]
    }
   ],
   "source": [
    "# Aminoácidos Pfeature\n",
    "aa_pfeature = ['A','C','D','E','F','G','H','I','K','L','M','N','P',\n",
    "              'Q','R','S','T','V','W','Y']\n",
    "\n",
    "# Aminoácidos IUPAC - Pfeatures\n",
    "aa_iupac = ['B','U','X','Z']\n",
    "\n",
    "# Aminoácidos en minúsculas\n",
    "aa_min = [elemento.lower() for elemento in (aa_pfeature + aa_iupac)]\n",
    "\n",
    "print ('Total de aminoácidos IUPAC:', len(aa_iupac) + len(aa_pfeature))\n",
    "print ('Total de aminoácidos Pfeatures:', len(aa_pfeature))\n",
    "print ('Total de aminoácidos que están en UIPAC pero no reconoce Pfeatures:', len(aa_iupac))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres únicos:  64\n"
     ]
    }
   ],
   "source": [
    "# Crear un diccionario para contar la aparición de cada carácter\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in df_peptidos['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / df_peptidos.shape[0]).round(4) *100\n",
    "\n",
    "\n",
    "print('Total de caracteres únicos: ', len(char_count_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>7</th>\n",
       "      <th>6</th>\n",
       "      <th>2</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>4</th>\n",
       "      <th>12</th>\n",
       "      <th>5</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>10</th>\n",
       "      <th>17</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>9</th>\n",
       "      <th>19</th>\n",
       "      <th>18</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>L</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>F</td>\n",
       "      <td>P</td>\n",
       "      <td>W</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>3858</td>\n",
       "      <td>3390</td>\n",
       "      <td>3207</td>\n",
       "      <td>3081</td>\n",
       "      <td>3014</td>\n",
       "      <td>2964</td>\n",
       "      <td>2932</td>\n",
       "      <td>2875</td>\n",
       "      <td>2435</td>\n",
       "      <td>2143</td>\n",
       "      <td>2077</td>\n",
       "      <td>1985</td>\n",
       "      <td>1942</td>\n",
       "      <td>1825</td>\n",
       "      <td>1824</td>\n",
       "      <td>1808</td>\n",
       "      <td>1684</td>\n",
       "      <td>1202</td>\n",
       "      <td>1116</td>\n",
       "      <td>1033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>65.94</td>\n",
       "      <td>57.94</td>\n",
       "      <td>54.81</td>\n",
       "      <td>52.66</td>\n",
       "      <td>51.51</td>\n",
       "      <td>50.66</td>\n",
       "      <td>50.11</td>\n",
       "      <td>49.14</td>\n",
       "      <td>41.62</td>\n",
       "      <td>36.63</td>\n",
       "      <td>35.5</td>\n",
       "      <td>33.93</td>\n",
       "      <td>33.19</td>\n",
       "      <td>31.19</td>\n",
       "      <td>31.17</td>\n",
       "      <td>30.9</td>\n",
       "      <td>28.78</td>\n",
       "      <td>20.54</td>\n",
       "      <td>19.07</td>\n",
       "      <td>17.66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              3      7      6      2      0      1      4      12     5   \\\n",
       "Character      L      K      R      I      A      G      V      F      P   \n",
       "Count       3858   3390   3207   3081   3014   2964   2932   2875   2435   \n",
       "%          65.94  57.94  54.81  52.66  51.51  50.66  50.11  49.14  41.62   \n",
       "\n",
       "              11    13     10     17     14     15    16     9      19     18  \\\n",
       "Character      W     S      Y      N      D      E     Q      T      M      C   \n",
       "Count       2143  2077   1985   1942   1825   1824  1808   1684   1202   1116   \n",
       "%          36.63  35.5  33.93  33.19  31.19  31.17  30.9  28.78  20.54  19.07   \n",
       "\n",
       "              8   \n",
       "Character      H  \n",
       "Count       1033  \n",
       "%          17.66  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos Pfeatures Mayúsculas\n",
    "char_count_aapf = char_count_df[char_count_df['Character'].isin(aa_pfeature)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aapf))\n",
    "char_count_aapf.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>29</th>\n",
       "      <th>51</th>\n",
       "      <th>37</th>\n",
       "      <th>41</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>X</td>\n",
       "      <td>Z</td>\n",
       "      <td>B</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>691</td>\n",
       "      <td>364</td>\n",
       "      <td>31</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>11.81</td>\n",
       "      <td>6.22</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              29    51    37    41\n",
       "Character      X     Z     B     U\n",
       "Count        691   364    31     5\n",
       "%          11.81  6.22  0.53  0.09"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos UIPAC no incluídos en Pfeatures Mayúsculas\n",
    "char_count_aauipac = char_count_df[char_count_df['Character'].isin(aa_iupac)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aauipac))\n",
    "char_count_aauipac.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La ocurrencia de Z, B, U es muy baja por lo que se procederá a eliminar todos los péptidos que las contienen. \n",
    "Para el caso de los péptidos con 'x' se re-escribiran aquellas que solo tengan X al inicio o al final. Por ejemplo, si la secuencia es XXVKKGLISY reescribirla como VKKGLISY, si la secuencia es KKGLISYXXX entonces se reescribirá como KKGLISY. Sin embargo, si la secuencia es KKGXXLISY entonces se eliminará la secuencia del dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres de Pfeatures incluidos en el dataset:  24\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>30</th>\n",
       "      <th>23</th>\n",
       "      <th>25</th>\n",
       "      <th>20</th>\n",
       "      <th>39</th>\n",
       "      <th>21</th>\n",
       "      <th>28</th>\n",
       "      <th>31</th>\n",
       "      <th>27</th>\n",
       "      <th>34</th>\n",
       "      <th>...</th>\n",
       "      <th>47</th>\n",
       "      <th>26</th>\n",
       "      <th>48</th>\n",
       "      <th>43</th>\n",
       "      <th>33</th>\n",
       "      <th>49</th>\n",
       "      <th>24</th>\n",
       "      <th>44</th>\n",
       "      <th>53</th>\n",
       "      <th>59</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>r</td>\n",
       "      <td>l</td>\n",
       "      <td>k</td>\n",
       "      <td>f</td>\n",
       "      <td>x</td>\n",
       "      <td>a</td>\n",
       "      <td>v</td>\n",
       "      <td>w</td>\n",
       "      <td>i</td>\n",
       "      <td>n</td>\n",
       "      <td>...</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>q</td>\n",
       "      <td>e</td>\n",
       "      <td>m</td>\n",
       "      <td>h</td>\n",
       "      <td>d</td>\n",
       "      <td>u</td>\n",
       "      <td>z</td>\n",
       "      <td>b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>141</td>\n",
       "      <td>132</td>\n",
       "      <td>112</td>\n",
       "      <td>78</td>\n",
       "      <td>77</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>52</td>\n",
       "      <td>44</td>\n",
       "      <td>33</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>2.41</td>\n",
       "      <td>2.26</td>\n",
       "      <td>1.91</td>\n",
       "      <td>1.33</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.56</td>\n",
       "      <td>...</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             30    23    25    20    39    21    28    31    27    34  ...  \\\n",
       "Character     r     l     k     f     x     a     v     w     i     n  ...   \n",
       "Count       141   132   112    78    77    54    54    52    44    33  ...   \n",
       "%          2.41  2.26  1.91  1.33  1.32  0.92  0.92  0.89  0.75  0.56  ...   \n",
       "\n",
       "             47    26    48    43    33    49    24    44    53    59  \n",
       "Character     c     g     q     e     m     h     d     u     z     b  \n",
       "Count        21    19    15    13    11     8     7     1     1     1  \n",
       "%          0.36  0.32  0.26  0.22  0.19  0.14  0.12  0.02  0.02  0.02  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aminoácidos UIPAC no incluídos en Pfeatures Mayúsculas\n",
    "char_count_aamin = char_count_df[char_count_df['Character'].isin(aa_min)]\n",
    "print('Total de caracteres de Pfeatures incluidos en el dataset: ', len(char_count_aamin))\n",
    "char_count_aamin.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La distincion entre los tipos de aminoácidos L o sinteticos y D o naturales se puede ver en estos resultados, siendo representados los tipo D con letras mayusculuas y los tipo L con minusculas. Gracias a este analisis podemos mostrar que no es relevante mantener la distincion entre estos aminoácidos y se hara un tratamiento a estos datos para convertir los caracteres en minusculas a mayusculas mas adelante.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres no incluidos en el UIPAC:  16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>46</th>\n",
       "      <th>52</th>\n",
       "      <th>54</th>\n",
       "      <th>32</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>45</th>\n",
       "      <th>50</th>\n",
       "      <th>42</th>\n",
       "      <th>38</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Character</th>\n",
       "      <td>Ⓧ</td>\n",
       "      <td>,</td>\n",
       "      <td>Φ</td>\n",
       "      <td>O</td>\n",
       "      <td>(</td>\n",
       "      <td>)</td>\n",
       "      <td>-</td>\n",
       "      <td>j</td>\n",
       "      <td></td>\n",
       "      <td>工</td>\n",
       "      <td>]</td>\n",
       "      <td>[</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Count</th>\n",
       "      <td>236</td>\n",
       "      <td>18</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>%</th>\n",
       "      <td>4.03</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             46    52    54    32    55    56    45    50    42    38    57  \\\n",
       "Character     Ⓧ     ,     Φ     O     (     )     -     j           工     ]   \n",
       "Count       236    18    11    10     4     4     3     3     2     1     1   \n",
       "%          4.03  0.31  0.19  0.17  0.07  0.07  0.05  0.05  0.03  0.02  0.02   \n",
       "\n",
       "             58    60    61    62    63  \n",
       "Character     [     6     2     1     7  \n",
       "Count         1     1     1     1     1  \n",
       "%          0.02  0.02  0.02  0.02  0.02  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Número de péptidos en los que está cada caracter que no es aminoácido IUPAC:\n",
    "char_count_otros = char_count_df[~char_count_df['Character'].isin(aa_pfeature + aa_iupac + aa_min)]\n",
    "print('Total de caracteres no incluidos en el UIPAC: ', len(char_count_otros))\n",
    "char_count_otros.sort_values(by='Count', ascending=False).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['O', '工', '\\xa0', '-', 'Ⓧ', 'j', ',', 'Φ', '(', ')', ']', '[', '6',\n",
       "       '2', '1', '7'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "char_count_otros['Character'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La frecuencia de los caractéres no incluidos en IUPAC es muy baja, por lo que también se eliminarán los péptidos que las contienen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ajuste de las secuencias de aminoácidos\n",
    "filtered_df = df_peptidos.copy()\n",
    "# 1. Poner en mayúsculas todos los aminoácidos de las secuencias\n",
    "filtered_df['Sequence'] = filtered_df['Sequence'].str.upper()\n",
    "# 2. Identificar los péptidos con valores de X\n",
    "aa_x = filtered_df[filtered_df['Sequence'].apply(lambda x: 'X' in x)]\n",
    "# 3. Conservar solo los péptidos que contengan los aminoácidos para pfeatures\n",
    "filtered_df = filtered_df[filtered_df['Sequence'].apply(lambda x: all(char in aa_pfeature for char in x))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de péptidos con X 727\n",
      "Péptidos que empiezan con X:  257\n",
      "Péptidos que finalizan con X:  177\n",
      "Péptidos con X intermedia: 551\n",
      "Peptidos conservados:  176\n"
     ]
    }
   ],
   "source": [
    "# Identifica todos péptidos con valores de X\n",
    "total_x = len(aa_x)\n",
    "print('Total de péptidos con X', total_x)\n",
    "\n",
    "# Inician con X \n",
    "inicio_x = aa_x[aa_x['Sequence'].str.startswith('X')]\n",
    "print('Péptidos que empiezan con X: ', len(inicio_x))\n",
    "# Función para quitar los caracteres 'x' al inicio de la secuencia\n",
    "def quitar_x_inicio(secuencia):\n",
    "    return re.sub(r'^X+', '', secuencia)\n",
    "# Aplicar la función a la columna 'Sequence'\n",
    "aa_x['Sequence'] = aa_x['Sequence'].apply(quitar_x_inicio)\n",
    "\n",
    "# Finalizan con X\n",
    "fin_x = aa_x[aa_x['Sequence'].str.endswith('X')].copy()\n",
    "print('Péptidos que finalizan con X: ', len(fin_x))\n",
    "# Función para quitar los caracteres 'x' al final de la secuencia\n",
    "def quitar_x_final(secuencia):\n",
    "    return re.sub(r'X+$', '', secuencia)\n",
    "# Aplicar la función a la columna 'Sequence'\n",
    "aa_x['Sequence'] = aa_x['Sequence'].apply(quitar_x_final)\n",
    "\n",
    "# Elimina los péptidos con X intermedia\n",
    "aa_x = aa_x[aa_x['Sequence'].apply(lambda x: all(char in aa_pfeature for char in x))]\n",
    "print('Péptidos con X intermedia:', total_x - len(aa_x))\n",
    "print('Peptidos conservados: ', len(aa_x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de péptidos:  5851\n",
      "Péptidos solo con aminoácidos para pFeatures:  4679\n",
      "Péptidos eliminados:  20.03 %\n"
     ]
    }
   ],
   "source": [
    "# Agrega los peptidos con X modificados al dataframe de péptidos depurados\n",
    "filtered_df = pd.concat([filtered_df, aa_x], ignore_index=True)\n",
    "\n",
    "print('Total de péptidos: ', df_peptidos.shape[0])\n",
    "print('Péptidos solo con aminoácidos para pFeatures: ', filtered_df.shape[0])\n",
    "print('Péptidos eliminados: ', round((df_peptidos.shape[0]-filtered_df.shape[0])/df_peptidos.shape[0],4) * 100, '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de caracteres únicos:  20\n"
     ]
    }
   ],
   "source": [
    "# Valida los aminoácidos contenidos\n",
    "char_count = {}\n",
    "\n",
    "# Iterar sobre cada secuencia en la columna 'Sequence'\n",
    "for sequence in filtered_df['Sequence']:\n",
    "    # Convertir la secuencia en un conjunto de caracteres únicos\n",
    "    unique_chars = set(sequence)\n",
    "    # Contar la aparición de cada carácter en el conjunto\n",
    "    for char in unique_chars:\n",
    "        if char in char_count:\n",
    "            char_count[char] += 1\n",
    "        else:\n",
    "            char_count[char] = 1\n",
    "\n",
    "# Convertir el diccionario en un DataFrame para una mejor visualización\n",
    "char_count_df = pd.DataFrame(list(char_count.items()), columns=['Character', 'Count'])\n",
    "char_count_df['%'] = (char_count_df['Count'] / filtered_df.shape[0]).round(4) *100\n",
    "\n",
    "\n",
    "print('Total de caracteres únicos: ', len(char_count_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df.to_csv('../data/processed/peptide_filtered.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convertir dataframe to FASTA \n",
    "\n",
    "Para los siguientes pasos del proceso, es necesario hacer la conversion de los datos filtrados a formato FASTA, este formato es el estandard usado para las secuencias de aminoácidos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('../data/processed/VEGF.fasta','w', encoding='UTF8') as f:\n",
    "    for _, row in filtered_df.iterrows():\n",
    "        f.write(f\">{row['VEGF']}\\n{row['Sequence']}\\n\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ejecutar CD-HIT\n",
    "\n",
    "CD-HIT es un programa ampliamente utilizado para agrupar secuencias biológicas con el fin de reducir la redundancia de secuencias y mejorar el rendimiento de otros análisis de secuencias. En respuesta al rápido aumento de la cantidad de datos de secuenciación producidos por las tecnologías de secuenciación de próxima generación, hemos desarrollado un nuevo programa CD-HIT acelerado con una novedosa estrategia de paralelización y algunas otras técnicas para permitir la agrupación eficiente de dichos conjuntos de datos. Nuestras pruebas demostraron una aceleración muy buena derivada de la paralelización para hasta ∼24 núcleos y una aceleración cuasi-lineal para hasta ∼8 núcleos. El CD-HIT mejorado es capaz de manejar conjuntos de datos muy grandes en mucho menos tiempo que las versiones anteriores[2]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Aug 20 2021, 08:39:56\n",
      "Command: cd-hit -i ../data/processed/VEGF.fasta -o\n",
      "         ../data/processed/VEGF.txt -c 0.99\n",
      "\n",
      "Started: Sun Feb  2 18:14:55 2025\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 3276\n",
      "longest and shortest : 97 and 11\n",
      "Total letters: 68527\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 76M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 4000000\n",
      "Max number of word counting entries: 90445565\n",
      "\n",
      "\n",
      "comparing sequences from          0  to       3276\n",
      "...\n",
      "     3276  finished       2704  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.17\n",
      "\n",
      "STDERR: w\u0000s\u0000l\u0000:\u0000 \u0000A\u0000 \u0000l\u0000o\u0000c\u0000a\u0000l\u0000h\u0000o\u0000s\u0000t\u0000 \u0000p\u0000r\u0000o\u0000x\u0000y\u0000 \u0000c\u0000o\u0000n\u0000f\u0000i\u0000g\u0000u\u0000r\u0000a\u0000t\u0000i\u0000o\u0000n\u0000 \u0000w\u0000a\u0000s\u0000 \u0000d\u0000e\u0000t\u0000e\u0000c\u0000t\u0000e\u0000d\u0000 \u0000b\u0000u\u0000t\u0000 \u0000n\u0000o\u0000t\u0000 \u0000m\u0000i\u0000r\u0000r\u0000o\u0000r\u0000e\u0000d\u0000 \u0000i\u0000n\u0000t\u0000o\u0000 \u0000W\u0000S\u0000L\u0000.\u0000 \u0000W\u0000S\u0000L\u0000 \u0000i\u0000n\u0000 \u0000N\u0000A\u0000T\u0000 \u0000m\u0000o\u0000d\u0000e\u0000 \u0000d\u0000o\u0000e\u0000s\u0000 \u0000n\u0000o\u0000t\u0000 \u0000s\u0000u\u0000p\u0000p\u0000o\u0000r\u0000t\u0000 \u0000l\u0000o\u0000c\u0000a\u0000l\u0000h\u0000o\u0000s\u0000t\u0000 \u0000p\u0000r\u0000o\u0000x\u0000i\u0000e\u0000s\u0000.\u0000\n",
      "\u0000\n",
      "\u0000\n"
     ]
    }
   ],
   "source": [
    "#Ejecutar cd-hit en colab, para windows usar el comando con wsl\n",
    "#! cd-hit -i data/processed/VEGF.fasta -o data/processed/VEGF.txt -c 0.99\n",
    "\n",
    "import subprocess\n",
    "\n",
    "# Define the WSL command buscar \n",
    "command = [\"wsl\", \"cd-hit\", \"-i\", \"../data/processed/VEGF.fasta\", \"-o\", \"../data/processed/VEGF.txt\", \"-c\", \"0.99\"]\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(command, capture_output=True, text=True)\n",
    "\n",
    "# Print the output and errors\n",
    "print(\"STDOUT:\", result.stdout)\n",
    "print(\"STDERR:\", result.stderr)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Funciones de Pfeature\n",
    "\n",
    "La librería pfeature es una herramienta de Python utilizada para extraer y analizar características de secuencias proteicas y nucleotídicas. Proporciona métodos para calcular descriptores físicos, químicos, estadísticos y estructurales de biomoléculas, lo que la hace útil en bioinformática, predicción de propiedades funcionales y clasificación de secuencias en aprendizaje automático.\n",
    "\n",
    "\n",
    "## Aplicaciones en Bioinformática\n",
    "pfeature se usa ampliamente en:\n",
    "- Predicción de propiedades de proteínas (hidrofobicidad, accesibilidad al solvente, estabilidad).\n",
    "- Clasificación de proteínas en modelos de Machine Learning.\n",
    "- Análisis de secuencias en estudios evolutivos.\n",
    "- Desarrollo de modelos para predicción de interacciones proteína-ligando.\n",
    "\n",
    "pfeature es una poderosa herramienta en bioinformática para analizar secuencias biológicas de manera rápida y eficiente. Su capacidad de generar múltiples descriptores permite su aplicación en aprendizaje automático, descubrimiento de fármacos y análisis estructural de proteínas[3].\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lee el archivo creado por CD-HIT y almacena las etiquetas en un archivo CSV para su uso posterior\n",
    "cd_hit_path = \"../data/processed/VEGF.txt\"\n",
    "records_cd_hit = list(SeqIO.parse(cd_hit_path, \"fasta\"))\n",
    "with open('../data/processed/VEGF_y.csv','w', encoding='UTF8') as f:\n",
    "    for record in records_cd_hit:\n",
    "        f.write(f'{record.id} \\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Physico-Chemical Properties Composition: 30 Descriptores\n",
    "\n",
    "Este modulo calcula la fracción de cada propiedas físico-química de una secuencia. Las siguientes propiedades se han incorporado a Pfeature para calcular caracteristicas composicionales basadas en propiedades físico-químicas estandard.\n",
    "\n",
    "- Positively Charged → Carga Positiva\n",
    "- Negatively Charged → Carga Negativa\n",
    "- Neutral Charge → Carga Neutra\n",
    "- Polarity in residues → Polaridad en residuos\n",
    "- Non-polarity in residues → No polaridad en residuos\n",
    "- Aliphaticity → Alifaticidad\n",
    "- Cyclicity → Ciclicidad\n",
    "- Aromaticity → Aromaticidad\n",
    "- Acidity → Acidez\n",
    "- Basicity → Basicidad\n",
    "- Neutral (pH) → Neutro (pH)\n",
    "- Hydrophobicity → Hidrofobicidad\n",
    "- Hydrophilicity → Hidrofilicidad\n",
    "- Neutral towards water → Neutro frente al agua\n",
    "- Hydroxylic → Hidroxílico\n",
    "- Sulphur Content → Contenido de Azufre\n",
    "- Tiny → Pequeño\n",
    "- Small → Mediano\n",
    "- Large → Grande\n",
    "\n",
    "parfa el cálculo de estas propiedades se usa la siguiente formula \n",
    "\n",
    "### 𝐶𝑃𝑖 = 𝑃𝑖/𝐿 \n",
    "\n",
    "donde\n",
    "\n",
    "- 𝐶𝑃𝑖 = Propiedades físico-quimicas de la composicion de tipo 𝑖.\n",
    "- 𝑃𝑖 = suma de la propiedad de tipo 𝑖.\n",
    "- 𝐿 = Longitud de la secuencia. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCP_PC</th>\n",
       "      <th>PCP_NC</th>\n",
       "      <th>PCP_NE</th>\n",
       "      <th>PCP_PO</th>\n",
       "      <th>PCP_NP</th>\n",
       "      <th>PCP_AL</th>\n",
       "      <th>PCP_CY</th>\n",
       "      <th>PCP_AR</th>\n",
       "      <th>PCP_AC</th>\n",
       "      <th>PCP_BS</th>\n",
       "      <th>...</th>\n",
       "      <th>PCP_SA_EX</th>\n",
       "      <th>PCP_SA_IN</th>\n",
       "      <th>PCP_TN</th>\n",
       "      <th>PCP_SM</th>\n",
       "      <th>PCP_LR</th>\n",
       "      <th>PCP_Z1</th>\n",
       "      <th>PCP_Z2</th>\n",
       "      <th>PCP_Z3</th>\n",
       "      <th>PCP_Z4</th>\n",
       "      <th>PCP_Z5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.357</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.357</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.357</td>\n",
       "      <td>0.643</td>\n",
       "      <td>0.075</td>\n",
       "      <td>-0.601</td>\n",
       "      <td>-1.386</td>\n",
       "      <td>0.315</td>\n",
       "      <td>0.169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.800</td>\n",
       "      <td>-0.407</td>\n",
       "      <td>0.018</td>\n",
       "      <td>-0.867</td>\n",
       "      <td>0.655</td>\n",
       "      <td>0.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.750</td>\n",
       "      <td>-0.295</td>\n",
       "      <td>0.118</td>\n",
       "      <td>-0.689</td>\n",
       "      <td>0.732</td>\n",
       "      <td>0.198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.444</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.556</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.278</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.444</td>\n",
       "      <td>...</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.833</td>\n",
       "      <td>-0.556</td>\n",
       "      <td>0.583</td>\n",
       "      <td>-0.806</td>\n",
       "      <td>0.904</td>\n",
       "      <td>0.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.812</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.125</td>\n",
       "      <td>...</td>\n",
       "      <td>0.188</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.312</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.500</td>\n",
       "      <td>-0.768</td>\n",
       "      <td>-1.499</td>\n",
       "      <td>-0.647</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.280</td>\n",
       "      <td>0.320</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.480</td>\n",
       "      <td>1.878</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.327</td>\n",
       "      <td>0.854</td>\n",
       "      <td>-0.409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>0.412</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.588</td>\n",
       "      <td>0.176</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.412</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.118</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.471</td>\n",
       "      <td>1.835</td>\n",
       "      <td>-0.429</td>\n",
       "      <td>-0.168</td>\n",
       "      <td>1.029</td>\n",
       "      <td>-0.636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>0.444</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.556</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.444</td>\n",
       "      <td>...</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1.354</td>\n",
       "      <td>0.284</td>\n",
       "      <td>-1.157</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>0.909</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0.909</td>\n",
       "      <td>...</td>\n",
       "      <td>0.182</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>3.371</td>\n",
       "      <td>2.150</td>\n",
       "      <td>-3.100</td>\n",
       "      <td>1.487</td>\n",
       "      <td>-0.134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>0.167</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.833</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.167</td>\n",
       "      <td>...</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.267</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.567</td>\n",
       "      <td>-0.549</td>\n",
       "      <td>-1.084</td>\n",
       "      <td>-0.920</td>\n",
       "      <td>-0.202</td>\n",
       "      <td>0.240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PCP_PC  PCP_NC  PCP_NE  PCP_PO  PCP_NP  PCP_AL  PCP_CY  PCP_AR  PCP_AC  \\\n",
       "0      0.357   0.000   0.643   0.071   0.571   0.571   0.036   0.036   0.000   \n",
       "1      0.400   0.000   0.600   0.050   0.550   0.350   0.000   0.200   0.000   \n",
       "2      0.400   0.000   0.600   0.050   0.550   0.350   0.050   0.200   0.000   \n",
       "3      0.444   0.000   0.556   0.056   0.500   0.278   0.056   0.222   0.000   \n",
       "4      0.125   0.062   0.812   0.062   0.750   0.688   0.000   0.062   0.062   \n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "2699   0.400   0.000   0.600   0.280   0.320   0.240   0.000   0.080   0.000   \n",
       "2700   0.412   0.000   0.588   0.176   0.412   0.353   0.000   0.059   0.000   \n",
       "2701   0.444   0.000   0.556   0.167   0.389   0.333   0.111   0.056   0.000   \n",
       "2702   0.909   0.091   0.000   0.000   0.000   0.000   0.000   0.000   0.091   \n",
       "2703   0.167   0.000   0.833   0.167   0.667   0.600   0.033   0.067   0.000   \n",
       "\n",
       "      PCP_BS  ...  PCP_SA_EX  PCP_SA_IN  PCP_TN  PCP_SM  PCP_LR  PCP_Z1  \\\n",
       "0      0.357  ...      0.143      0.143   0.214   0.357   0.643   0.075   \n",
       "1      0.400  ...      0.350      0.100   0.200   0.200   0.800  -0.407   \n",
       "2      0.400  ...      0.400      0.150   0.200   0.250   0.750  -0.295   \n",
       "3      0.444  ...      0.444      0.167   0.111   0.167   0.833  -0.556   \n",
       "4      0.125  ...      0.188      0.062   0.312   0.500   0.500  -0.768   \n",
       "...      ...  ...        ...        ...     ...     ...     ...     ...   \n",
       "2699   0.400  ...      0.000      0.240   0.520   0.520   0.480   1.878   \n",
       "2700   0.412  ...      0.000      0.118   0.529   0.529   0.471   1.835   \n",
       "2701   0.444  ...      0.333      0.111   0.222   0.333   0.667   1.354   \n",
       "2702   0.909  ...      0.182      0.000   0.000   0.000   1.000   3.371   \n",
       "2703   0.167  ...      0.200      0.133   0.267   0.433   0.567  -0.549   \n",
       "\n",
       "      PCP_Z2  PCP_Z3  PCP_Z4  PCP_Z5  \n",
       "0     -0.601  -1.386   0.315   0.169  \n",
       "1      0.018  -0.867   0.655   0.111  \n",
       "2      0.118  -0.689   0.732   0.198  \n",
       "3      0.583  -0.806   0.904   0.262  \n",
       "4     -1.499  -0.647  -0.537   0.250  \n",
       "...      ...     ...     ...     ...  \n",
       "2699  -0.005  -0.327   0.854  -0.409  \n",
       "2700  -0.429  -0.168   1.029  -0.636  \n",
       "2701   0.284  -1.157   0.621   0.075  \n",
       "2702   2.150  -3.100   1.487  -0.134  \n",
       "2703  -1.084  -0.920  -0.202   0.240  \n",
       "\n",
       "[2704 rows x 30 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Physico-Chemical Properties Composition (PCP)\n",
    "\n",
    "def pcp(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'dpc.csv'\n",
    "  df_out = pcp_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_pcp = pcp(cd_hit_path)\n",
    "feature_pcp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlacion: Descarta características redundantes con alta correlación entre sí."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCP_PC</th>\n",
       "      <th>PCP_NC</th>\n",
       "      <th>PCP_PO</th>\n",
       "      <th>PCP_NP</th>\n",
       "      <th>PCP_AL</th>\n",
       "      <th>PCP_CY</th>\n",
       "      <th>PCP_AR</th>\n",
       "      <th>PCP_HB</th>\n",
       "      <th>PCP_HL</th>\n",
       "      <th>PCP_NT</th>\n",
       "      <th>...</th>\n",
       "      <th>PCP_SA_BU</th>\n",
       "      <th>PCP_SA_EX</th>\n",
       "      <th>PCP_SA_IN</th>\n",
       "      <th>PCP_TN</th>\n",
       "      <th>PCP_SM</th>\n",
       "      <th>PCP_Z1</th>\n",
       "      <th>PCP_Z2</th>\n",
       "      <th>PCP_Z3</th>\n",
       "      <th>PCP_Z4</th>\n",
       "      <th>PCP_Z5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.357</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.179</td>\n",
       "      <td>...</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.143</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.357</td>\n",
       "      <td>0.075</td>\n",
       "      <td>-0.601</td>\n",
       "      <td>-1.386</td>\n",
       "      <td>0.315</td>\n",
       "      <td>0.169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.150</td>\n",
       "      <td>...</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.200</td>\n",
       "      <td>-0.407</td>\n",
       "      <td>0.018</td>\n",
       "      <td>-0.867</td>\n",
       "      <td>0.655</td>\n",
       "      <td>0.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.550</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.150</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.150</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.250</td>\n",
       "      <td>-0.295</td>\n",
       "      <td>0.118</td>\n",
       "      <td>-0.689</td>\n",
       "      <td>0.732</td>\n",
       "      <td>0.198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.444</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.278</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.056</td>\n",
       "      <td>...</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.167</td>\n",
       "      <td>-0.556</td>\n",
       "      <td>0.583</td>\n",
       "      <td>-0.806</td>\n",
       "      <td>0.904</td>\n",
       "      <td>0.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.562</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.750</td>\n",
       "      <td>0.188</td>\n",
       "      <td>0.062</td>\n",
       "      <td>0.312</td>\n",
       "      <td>0.500</td>\n",
       "      <td>-0.768</td>\n",
       "      <td>-1.499</td>\n",
       "      <td>-0.647</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>0.400</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.280</td>\n",
       "      <td>0.320</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.400</td>\n",
       "      <td>0.360</td>\n",
       "      <td>...</td>\n",
       "      <td>0.440</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.520</td>\n",
       "      <td>0.520</td>\n",
       "      <td>1.878</td>\n",
       "      <td>-0.005</td>\n",
       "      <td>-0.327</td>\n",
       "      <td>0.854</td>\n",
       "      <td>-0.409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>0.412</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.176</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.353</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.059</td>\n",
       "      <td>0.294</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.588</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.118</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.529</td>\n",
       "      <td>1.835</td>\n",
       "      <td>-0.429</td>\n",
       "      <td>-0.168</td>\n",
       "      <td>1.029</td>\n",
       "      <td>-0.636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>0.444</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.389</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.056</td>\n",
       "      <td>0.278</td>\n",
       "      <td>0.556</td>\n",
       "      <td>0.278</td>\n",
       "      <td>...</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.333</td>\n",
       "      <td>1.354</td>\n",
       "      <td>0.284</td>\n",
       "      <td>-1.157</td>\n",
       "      <td>0.621</td>\n",
       "      <td>0.075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>0.909</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.909</td>\n",
       "      <td>0.091</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.182</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3.371</td>\n",
       "      <td>2.150</td>\n",
       "      <td>-3.100</td>\n",
       "      <td>1.487</td>\n",
       "      <td>-0.134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>0.167</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.067</td>\n",
       "      <td>0.567</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.633</td>\n",
       "      <td>0.200</td>\n",
       "      <td>0.133</td>\n",
       "      <td>0.267</td>\n",
       "      <td>0.433</td>\n",
       "      <td>-0.549</td>\n",
       "      <td>-1.084</td>\n",
       "      <td>-0.920</td>\n",
       "      <td>-0.202</td>\n",
       "      <td>0.240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PCP_PC  PCP_NC  PCP_PO  PCP_NP  PCP_AL  PCP_CY  PCP_AR  PCP_HB  PCP_HL  \\\n",
       "0      0.357   0.000   0.071   0.571   0.571   0.036   0.036   0.464   0.393   \n",
       "1      0.400   0.000   0.050   0.550   0.350   0.000   0.200   0.450   0.400   \n",
       "2      0.400   0.000   0.050   0.550   0.350   0.050   0.200   0.450   0.450   \n",
       "3      0.444   0.000   0.056   0.500   0.278   0.056   0.222   0.500   0.500   \n",
       "4      0.125   0.062   0.062   0.750   0.688   0.000   0.062   0.562   0.125   \n",
       "...      ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "2699   0.400   0.000   0.280   0.320   0.240   0.000   0.080   0.240   0.400   \n",
       "2700   0.412   0.000   0.176   0.412   0.353   0.000   0.059   0.294   0.412   \n",
       "2701   0.444   0.000   0.167   0.389   0.333   0.111   0.056   0.278   0.556   \n",
       "2702   0.909   0.091   0.000   0.000   0.000   0.000   0.000   0.000   0.909   \n",
       "2703   0.167   0.000   0.167   0.667   0.600   0.033   0.067   0.567   0.200   \n",
       "\n",
       "      PCP_NT  ...  PCP_SA_BU  PCP_SA_EX  PCP_SA_IN  PCP_TN  PCP_SM  PCP_Z1  \\\n",
       "0      0.179  ...      0.536      0.143      0.143   0.214   0.357   0.075   \n",
       "1      0.150  ...      0.550      0.350      0.100   0.200   0.200  -0.407   \n",
       "2      0.150  ...      0.500      0.400      0.150   0.200   0.250  -0.295   \n",
       "3      0.056  ...      0.444      0.444      0.167   0.111   0.167  -0.556   \n",
       "4      0.312  ...      0.750      0.188      0.062   0.312   0.500  -0.768   \n",
       "...      ...  ...        ...        ...        ...     ...     ...     ...   \n",
       "2699   0.360  ...      0.440      0.000      0.240   0.520   0.520   1.878   \n",
       "2700   0.294  ...      0.588      0.000      0.118   0.529   0.529   1.835   \n",
       "2701   0.278  ...      0.333      0.333      0.111   0.222   0.333   1.354   \n",
       "2702   0.091  ...      0.000      0.182      0.000   0.000   0.000   3.371   \n",
       "2703   0.333  ...      0.633      0.200      0.133   0.267   0.433  -0.549   \n",
       "\n",
       "      PCP_Z2  PCP_Z3  PCP_Z4  PCP_Z5  \n",
       "0     -0.601  -1.386   0.315   0.169  \n",
       "1      0.018  -0.867   0.655   0.111  \n",
       "2      0.118  -0.689   0.732   0.198  \n",
       "3      0.583  -0.806   0.904   0.262  \n",
       "4     -1.499  -0.647  -0.537   0.250  \n",
       "...      ...     ...     ...     ...  \n",
       "2699  -0.005  -0.327   0.854  -0.409  \n",
       "2700  -0.429  -0.168   1.029  -0.636  \n",
       "2701   0.284  -1.157   0.621   0.075  \n",
       "2702   2.150  -3.100   1.487  -0.134  \n",
       "2703  -1.084  -0.920  -0.202   0.240  \n",
       "\n",
       "[2704 rows x 25 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#correlacion\n",
    "corr_matrix = pd.DataFrame(feature_pcp).corr().abs()\n",
    "upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(bool))\n",
    "to_drop = [column for column in upper.columns if any(upper[column] > 0.9)]\n",
    "corr_matrix_pcp = feature_pcp.drop(to_drop, axis=1)\n",
    "corr_matrix_pcp.to_csv('../data/processed/corr_matrix_pcp.csv', index = False)\n",
    "corr_matrix_pcp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Composition enhanced Transition and Distribution: 189 descriptors [4]\n",
    "\n",
    "Composition-Enhanced Transition and Distribution (CTD) es un enfoque utilizado en bioinformática y química computacional para describir secuencias de aminoácidos o nucleótidos. Se basa en tres principios clave:\n",
    "\n",
    "Composición (Composition): Representa la proporción de cada tipo de aminoácido o nucleótido en una secuencia.\n",
    "Transición (Transition): Describe la frecuencia con la que un tipo de aminoácido/nucleótido cambia a otro a lo largo de la secuencia.\n",
    "Distribución (Distribution): Indica la posición relativa de ciertos tipos de aminoácidos/nucleótidos dentro de la secuencia.\n",
    "\n",
    "Cómo funciona CTD en la descripción de secuencias\n",
    "\n",
    "1. Composición (C - Group 1) \n",
    "\n",
    "- Calcula el porcentaje de cada grupo de aminoácidos en la secuencia.\n",
    "- Se agrupan los aminoácidos según ciertas propiedades fisicoquímicas, como hidrofobicidad o carga.\n",
    "\n",
    "2. Transición (T - Group 2)\n",
    "\n",
    "- Representa la frecuencia de cambio entre dos grupos de aminoácidos consecutivos en la secuencia.\n",
    "- Se calcula como el número de transiciones entre los grupos dividido por el total de transiciones posibles.\n",
    "\n",
    "3. Distribución (D - Group 3)\n",
    "\n",
    "- Indica en qué posiciones de la secuencia aparecen los primeros, el 25%, 50%, 75% y 100% de los residuos de un grupo específico.\n",
    "\n",
    "|Attributes|Group1|Group 2|Group 3|\n",
    "|:-:|:-:|:-:|:-:|\n",
    "|Hydrophobicity| R,K,E,D,Q,N| G,A,S,T,P,H,Y| C,L,V,I,M,F,W|\n",
    "|Normalized Vander Waals volume| G,A,S,T,P,D| N,V,E,Q,I,L| M,H,K,F,R,Y,W|\n",
    "|Polarity| L,I,F,W,C,M,V,Y| P,A,T,G,S| H,Q,R,K,N,E,D|\n",
    "|Polarizability| G,A,S,D,T| C,P,N,V,E,Q,I,L| K,M,H,F,R,Y,W|\n",
    "|Charge| K,R| A,N,C,Q,G,H,I,L,M,F,P,S,T,W,Y,V| D,E|\n",
    "|Secondary structure| E,A,L,M,Q,K,R,H| V,I,Y,C,W,F,T| G,N,P,S,D|\n",
    "|Solvent accessibility| A,L,F,C,G,I,V,W| R,K,Q,E,N,D| M,S P,T,H,Y|\n",
    "\n",
    "Los atributos de CTD permiten capturar información estructural y funcional de secuencias biológicas, convirtiéndolas en vectores numéricos que pueden ser utilizados para la predicción y clasificación en bioinformática."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CeTD_HB1</th>\n",
       "      <th>CeTD_HB2</th>\n",
       "      <th>CeTD_HB3</th>\n",
       "      <th>CeTD_VW1</th>\n",
       "      <th>CeTD_VW2</th>\n",
       "      <th>CeTD_VW3</th>\n",
       "      <th>CeTD_PO1</th>\n",
       "      <th>CeTD_PO2</th>\n",
       "      <th>CeTD_PO3</th>\n",
       "      <th>CeTD_PZ1</th>\n",
       "      <th>...</th>\n",
       "      <th>CeTD_0_p_SS3</th>\n",
       "      <th>CeTD_25_p_SS3</th>\n",
       "      <th>CeTD_50_p_SS3</th>\n",
       "      <th>CeTD_75_p_SS3</th>\n",
       "      <th>CeTD_100_p_SS3</th>\n",
       "      <th>CeTD_0_p_SA3</th>\n",
       "      <th>CeTD_25_p_SA3</th>\n",
       "      <th>CeTD_50_p_SA3</th>\n",
       "      <th>CeTD_75_p_SA3</th>\n",
       "      <th>CeTD_100_p_SA3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32.14</td>\n",
       "      <td>35.71</td>\n",
       "      <td>32.14</td>\n",
       "      <td>28.57</td>\n",
       "      <td>32.14</td>\n",
       "      <td>39.29</td>\n",
       "      <td>35.71</td>\n",
       "      <td>28.57</td>\n",
       "      <td>35.71</td>\n",
       "      <td>25.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>60.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>35.00</td>\n",
       "      <td>30.00</td>\n",
       "      <td>35.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>15.00</td>\n",
       "      <td>60.00</td>\n",
       "      <td>35.00</td>\n",
       "      <td>25.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>38.89</td>\n",
       "      <td>22.22</td>\n",
       "      <td>38.89</td>\n",
       "      <td>16.67</td>\n",
       "      <td>16.67</td>\n",
       "      <td>66.67</td>\n",
       "      <td>38.89</td>\n",
       "      <td>16.67</td>\n",
       "      <td>44.44</td>\n",
       "      <td>11.11</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18.75</td>\n",
       "      <td>31.25</td>\n",
       "      <td>50.00</td>\n",
       "      <td>37.50</td>\n",
       "      <td>43.75</td>\n",
       "      <td>18.75</td>\n",
       "      <td>50.00</td>\n",
       "      <td>31.25</td>\n",
       "      <td>18.75</td>\n",
       "      <td>37.50</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>32.00</td>\n",
       "      <td>48.00</td>\n",
       "      <td>20.00</td>\n",
       "      <td>45.45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>54.55</td>\n",
       "      <td>20.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>40.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>29.41</td>\n",
       "      <td>47.06</td>\n",
       "      <td>23.53</td>\n",
       "      <td>42.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>57.14</td>\n",
       "      <td>23.53</td>\n",
       "      <td>35.29</td>\n",
       "      <td>41.18</td>\n",
       "      <td>35.29</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>55.56</td>\n",
       "      <td>27.78</td>\n",
       "      <td>16.67</td>\n",
       "      <td>29.41</td>\n",
       "      <td>17.65</td>\n",
       "      <td>52.94</td>\n",
       "      <td>16.67</td>\n",
       "      <td>27.78</td>\n",
       "      <td>55.56</td>\n",
       "      <td>16.67</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.09</td>\n",
       "      <td>90.91</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>23.33</td>\n",
       "      <td>36.67</td>\n",
       "      <td>40.00</td>\n",
       "      <td>36.67</td>\n",
       "      <td>40.00</td>\n",
       "      <td>23.33</td>\n",
       "      <td>40.00</td>\n",
       "      <td>36.67</td>\n",
       "      <td>23.33</td>\n",
       "      <td>33.33</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 189 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CeTD_HB1  CeTD_HB2  CeTD_HB3  CeTD_VW1  CeTD_VW2  CeTD_VW3  CeTD_PO1  \\\n",
       "0        32.14     35.71     32.14     28.57     32.14     39.29     35.71   \n",
       "1        35.00     25.00     40.00     20.00     20.00     60.00     40.00   \n",
       "2        35.00     30.00     35.00     25.00     15.00     60.00     35.00   \n",
       "3        38.89     22.22     38.89     16.67     16.67     66.67     38.89   \n",
       "4        18.75     31.25     50.00     37.50     43.75     18.75     50.00   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "2699     32.00     48.00     20.00     45.45      0.00     54.55     20.00   \n",
       "2700     29.41     47.06     23.53     42.86      0.00     57.14     23.53   \n",
       "2701     55.56     27.78     16.67     29.41     17.65     52.94     16.67   \n",
       "2702    100.00      0.00      0.00      0.00      9.09     90.91      0.00   \n",
       "2703     23.33     36.67     40.00     36.67     40.00     23.33     40.00   \n",
       "\n",
       "      CeTD_PO2  CeTD_PO3  CeTD_PZ1  ...  CeTD_0_p_SS3  CeTD_25_p_SS3  \\\n",
       "0        28.57     35.71     25.00  ...             0              2   \n",
       "1        20.00     40.00     20.00  ...             0              1   \n",
       "2        25.00     40.00     20.00  ...             0              1   \n",
       "3        16.67     44.44     11.11  ...             0              1   \n",
       "4        31.25     18.75     37.50  ...             0              0   \n",
       "...        ...       ...       ...  ...           ...            ...   \n",
       "2699     40.00     40.00     40.00  ...             0              2   \n",
       "2700     35.29     41.18     35.29  ...             0              1   \n",
       "2701     27.78     55.56     16.67  ...             0              2   \n",
       "2702      0.00    100.00      0.00  ...             0              2   \n",
       "2703     36.67     23.33     33.33  ...             0              1   \n",
       "\n",
       "      CeTD_50_p_SS3  CeTD_75_p_SS3  CeTD_100_p_SS3  CeTD_0_p_SA3  \\\n",
       "0                 4              6               9             0   \n",
       "1                 3              5               7             0   \n",
       "2                 3              5               7             0   \n",
       "3                 3              5               7             0   \n",
       "4                 1              2               3             0   \n",
       "...             ...            ...             ...           ...   \n",
       "2699              4              6               8             0   \n",
       "2700              2              3               5             0   \n",
       "2701              5              7              10             0   \n",
       "2702              5              8              11             0   \n",
       "2703              3              5               7             0   \n",
       "\n",
       "      CeTD_25_p_SA3  CeTD_50_p_SA3  CeTD_75_p_SA3  CeTD_100_p_SA3  \n",
       "0                 1              2              3               4  \n",
       "1                 0              1              1               2  \n",
       "2                 0              1              2               3  \n",
       "3                 0              1              2               3  \n",
       "4                 0              0              0               1  \n",
       "...             ...            ...            ...             ...  \n",
       "2699              1              3              4               6  \n",
       "2700              0              1              1               2  \n",
       "2701              0              1              1               2  \n",
       "2702              0              0              0               0  \n",
       "2703              1              2              3               4  \n",
       "\n",
       "[2704 rows x 189 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Composition enhanced Transition and Distribution: (CTD) \n",
    "\n",
    "def ctd(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'ctd.csv'\n",
    "  df_out = ctd_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_ctd = ctd(cd_hit_path)\n",
    "feature_ctd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Se aplicara la transformación de PCA a las caracteristicas de CTD para reducir su dimensionalidad, manteniendo la mayor varianza posible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def aplicar_pca(data, threshold = 0.90):\n",
    "    print(\"Aplicando operación: PCA\")\n",
    "    scaler = StandardScaler()\n",
    "    X_scaled = scaler.fit_transform(data)\n",
    "    # Aplicar PCA\n",
    "    pca = PCA().fit(X_scaled)\n",
    "    # Calcular la varianza explicada acumulada\n",
    "    explained_variance = np.cumsum(pca.explained_variance_ratio_)\n",
    "    num_components = np.argmax(explained_variance >= threshold) + 1\n",
    "    pca = PCA(n_components=num_components)  # Número de componentes principales\n",
    "    X_pca = pca.fit_transform(X_scaled)\n",
    "    pca_columns = [f'PC{i+1}' for i in range(X_pca.shape[1])]  # Nombrar los componentes como PC1, PC2, etc.\n",
    "    df_pca = pd.DataFrame(X_pca, columns=pca_columns)\n",
    "    return df_pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>PC6</th>\n",
       "      <th>PC7</th>\n",
       "      <th>PC8</th>\n",
       "      <th>PC9</th>\n",
       "      <th>PC10</th>\n",
       "      <th>PC11</th>\n",
       "      <th>PC12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4.465813</td>\n",
       "      <td>2.238494</td>\n",
       "      <td>1.310017</td>\n",
       "      <td>-2.356987</td>\n",
       "      <td>1.922528</td>\n",
       "      <td>-0.451939</td>\n",
       "      <td>0.978216</td>\n",
       "      <td>-1.303886</td>\n",
       "      <td>-0.123767</td>\n",
       "      <td>0.780396</td>\n",
       "      <td>-1.447275</td>\n",
       "      <td>-0.078434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1.828824</td>\n",
       "      <td>3.707624</td>\n",
       "      <td>0.998182</td>\n",
       "      <td>-3.455248</td>\n",
       "      <td>1.169127</td>\n",
       "      <td>-0.739495</td>\n",
       "      <td>-0.578227</td>\n",
       "      <td>-0.117767</td>\n",
       "      <td>0.353232</td>\n",
       "      <td>-2.316752</td>\n",
       "      <td>-0.423856</td>\n",
       "      <td>0.288518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-2.042295</td>\n",
       "      <td>3.572486</td>\n",
       "      <td>-1.106014</td>\n",
       "      <td>-3.093488</td>\n",
       "      <td>0.903204</td>\n",
       "      <td>-1.017569</td>\n",
       "      <td>0.073823</td>\n",
       "      <td>-0.732423</td>\n",
       "      <td>1.842737</td>\n",
       "      <td>-2.697651</td>\n",
       "      <td>0.063404</td>\n",
       "      <td>0.224664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-3.946575</td>\n",
       "      <td>5.021183</td>\n",
       "      <td>0.124242</td>\n",
       "      <td>-3.025322</td>\n",
       "      <td>-0.214241</td>\n",
       "      <td>-1.147517</td>\n",
       "      <td>0.571092</td>\n",
       "      <td>0.412455</td>\n",
       "      <td>0.890656</td>\n",
       "      <td>-2.336612</td>\n",
       "      <td>-0.075183</td>\n",
       "      <td>0.033389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-3.182656</td>\n",
       "      <td>-6.694143</td>\n",
       "      <td>2.874705</td>\n",
       "      <td>0.816326</td>\n",
       "      <td>0.862016</td>\n",
       "      <td>0.666197</td>\n",
       "      <td>-2.222052</td>\n",
       "      <td>-0.948383</td>\n",
       "      <td>-0.788840</td>\n",
       "      <td>-0.248492</td>\n",
       "      <td>0.159610</td>\n",
       "      <td>0.657820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>2.191096</td>\n",
       "      <td>0.310381</td>\n",
       "      <td>-8.644810</td>\n",
       "      <td>-3.752692</td>\n",
       "      <td>1.559313</td>\n",
       "      <td>5.157465</td>\n",
       "      <td>-1.564594</td>\n",
       "      <td>-0.115044</td>\n",
       "      <td>-0.282376</td>\n",
       "      <td>-0.721614</td>\n",
       "      <td>-0.455279</td>\n",
       "      <td>2.394572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>-3.039970</td>\n",
       "      <td>-0.641169</td>\n",
       "      <td>-4.636086</td>\n",
       "      <td>-2.392446</td>\n",
       "      <td>2.341733</td>\n",
       "      <td>-1.132851</td>\n",
       "      <td>-1.104915</td>\n",
       "      <td>0.720910</td>\n",
       "      <td>-1.847882</td>\n",
       "      <td>-0.568111</td>\n",
       "      <td>0.674691</td>\n",
       "      <td>1.603461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>-4.126764</td>\n",
       "      <td>4.551046</td>\n",
       "      <td>-4.846577</td>\n",
       "      <td>1.338521</td>\n",
       "      <td>-0.069354</td>\n",
       "      <td>4.393361</td>\n",
       "      <td>0.463382</td>\n",
       "      <td>-1.405499</td>\n",
       "      <td>-1.271860</td>\n",
       "      <td>0.961104</td>\n",
       "      <td>0.422888</td>\n",
       "      <td>1.082130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>-10.472682</td>\n",
       "      <td>12.110773</td>\n",
       "      <td>-6.206620</td>\n",
       "      <td>2.676183</td>\n",
       "      <td>-1.859294</td>\n",
       "      <td>6.627050</td>\n",
       "      <td>-1.970334</td>\n",
       "      <td>-1.877130</td>\n",
       "      <td>-1.304498</td>\n",
       "      <td>0.329669</td>\n",
       "      <td>-0.330525</td>\n",
       "      <td>-1.962865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>6.380743</td>\n",
       "      <td>-4.548570</td>\n",
       "      <td>3.786818</td>\n",
       "      <td>-1.057695</td>\n",
       "      <td>2.421825</td>\n",
       "      <td>4.856283</td>\n",
       "      <td>-0.003916</td>\n",
       "      <td>-1.413393</td>\n",
       "      <td>0.771158</td>\n",
       "      <td>-0.268605</td>\n",
       "      <td>-0.543647</td>\n",
       "      <td>-0.011890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            PC1        PC2       PC3       PC4       PC5       PC6       PC7  \\\n",
       "0      4.465813   2.238494  1.310017 -2.356987  1.922528 -0.451939  0.978216   \n",
       "1     -1.828824   3.707624  0.998182 -3.455248  1.169127 -0.739495 -0.578227   \n",
       "2     -2.042295   3.572486 -1.106014 -3.093488  0.903204 -1.017569  0.073823   \n",
       "3     -3.946575   5.021183  0.124242 -3.025322 -0.214241 -1.147517  0.571092   \n",
       "4     -3.182656  -6.694143  2.874705  0.816326  0.862016  0.666197 -2.222052   \n",
       "...         ...        ...       ...       ...       ...       ...       ...   \n",
       "2699   2.191096   0.310381 -8.644810 -3.752692  1.559313  5.157465 -1.564594   \n",
       "2700  -3.039970  -0.641169 -4.636086 -2.392446  2.341733 -1.132851 -1.104915   \n",
       "2701  -4.126764   4.551046 -4.846577  1.338521 -0.069354  4.393361  0.463382   \n",
       "2702 -10.472682  12.110773 -6.206620  2.676183 -1.859294  6.627050 -1.970334   \n",
       "2703   6.380743  -4.548570  3.786818 -1.057695  2.421825  4.856283 -0.003916   \n",
       "\n",
       "           PC8       PC9      PC10      PC11      PC12  \n",
       "0    -1.303886 -0.123767  0.780396 -1.447275 -0.078434  \n",
       "1    -0.117767  0.353232 -2.316752 -0.423856  0.288518  \n",
       "2    -0.732423  1.842737 -2.697651  0.063404  0.224664  \n",
       "3     0.412455  0.890656 -2.336612 -0.075183  0.033389  \n",
       "4    -0.948383 -0.788840 -0.248492  0.159610  0.657820  \n",
       "...        ...       ...       ...       ...       ...  \n",
       "2699 -0.115044 -0.282376 -0.721614 -0.455279  2.394572  \n",
       "2700  0.720910 -1.847882 -0.568111  0.674691  1.603461  \n",
       "2701 -1.405499 -1.271860  0.961104  0.422888  1.082130  \n",
       "2702 -1.877130 -1.304498  0.329669 -0.330525 -1.962865  \n",
       "2703 -1.413393  0.771158 -0.268605 -0.543647 -0.011890  \n",
       "\n",
       "[2704 rows x 12 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_ctd = aplicar_pca(feature_ctd)\n",
    "pca_ctd.to_csv('../data/processed/pca_ctd.csv', index = False)\n",
    "pca_ctd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Entropia de Shannon\n",
    "\n",
    "La entropía de Shannon es una medida de la incertidumbre o diversidad en una secuencia de símbolos. En el contexto de los aminoácidos, se usa para cuantificar la variabilidad en la composición de una secuencia de proteínas o en una alineación de secuencias múltiples.\n",
    "\n",
    "### Fórmula de la Entropía de Shannon\n",
    "La entropía 𝐻 se define como:\n",
    "\n",
    "![alt text](image-1.png)\n",
    "\n",
    "​\n",
    "Donde:\n",
    "𝑝𝑖 es la frecuencia de aparición de un aminoácido 𝑖 en la secuencia.\n",
    "𝑛 es el número total de tipos de aminoácidos (generalmente 20).\n",
    "El logaritmo se toma en base 2 para obtener la entropía en bits.\n",
    "\n",
    "#Interpretación en Secuencias de Proteínas\n",
    "\n",
    "- Alta entropía: La secuencia tiene una composición muy diversa, sin un aminoácido predominante. Indica mayor variabilidad y menos conservación.\n",
    "- Baja entropía: Algunos aminoácidos dominan la secuencia, lo que sugiere regiones conservadas estructural o funcionalmente.\n",
    "\n",
    "### Aplicaciones en Bioinformática\n",
    "\n",
    "1. Análisis de Conservación: Se usa en alineamientos múltiples para identificar regiones conservadas en familias de proteínas.\n",
    "2. Predicción de Estructura y Función: Regiones con baja entropía suelen ser funcionalmente importantes (ej., sitios activos, dominios de unión).\n",
    "3. Estudios Evolutivos: Ayuda a detectar señales de selección natural en proteínas.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SEP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>2.549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>2.324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>2.711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>0.866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>3.483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        SEP\n",
       "0     3.200\n",
       "1     2.802\n",
       "2     2.902\n",
       "3     2.703\n",
       "4     3.031\n",
       "...     ...\n",
       "2699  2.549\n",
       "2700  2.324\n",
       "2701  2.711\n",
       "2702  0.866\n",
       "2703  3.483\n",
       "\n",
       "[2704 rows x 1 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def sep(input): #sShannon entropy\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'sep.csv'\n",
    "  df_out = sep_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_sep = sep(cd_hit_path)\n",
    "feature_sep.to_csv('../data/processed/Shannon_entropy.csv', index = False)\n",
    "feature_sep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Repetitive Residue Information (RRI)\n",
    "\n",
    "Repetitive Residue Information (RRI) es una métrica utilizada en bioinformática para analizar la redundancia y patrones de repetición en la secuencia de aminoácidos de péptidos y proteínas. Evalúa cuán frecuente es la aparición de residuos específicos en una secuencia y cómo esta repetición puede influir en la estructura y función de la biomolécula.\n",
    "\n",
    "### Definición de RRI\n",
    "RRI cuantifica la presencia de residuos repetitivos dentro de una secuencia peptídica. Se basa en la frecuencia de aparición de cada aminoácido y su distribución dentro de la secuencia. Esto permite identificar si una proteína o péptido tiene regiones altamente conservadas o patrones estructurales específicos, como repeticiones en hélices alfa o láminas beta.\n",
    "\n",
    "La fórmula general para calcular RRI es:\n",
    "\n",
    "![alt text](image.png)\n",
    "\n",
    "donde:\n",
    "- 𝑓𝑖 es la frecuencia del aminoácido 𝑖 en la secuencia.\n",
    "- 𝑁 es la longitud total de la secuencia.\n",
    "\n",
    "Esta ecuación es similar a la entropía de Shannon, midiendo la diversidad de residuos en la secuencia.\n",
    "\n",
    "### Aplicaciones de RRI en Bioinformática\n",
    "\n",
    "1. Identificación de motivos estructurales y funcionales\n",
    "\n",
    "- Péptidos con residuos repetitivos pueden formar estructuras secundarias estables, como hélices alfa anfipáticas o láminas beta repetitivas.\n",
    "\n",
    "2. Predicción de propiedades fisicoquímicas\n",
    "\n",
    "- Secuencias con alta repetición pueden afectar la hidrofobicidad, la solubilidad y la interacción con membranas celulares.\n",
    "\n",
    "3. Diseño de péptidos terapéuticos\n",
    "\n",
    "- Se utiliza en el diseño de péptidos antimicrobianos y en la optimización de fármacos basados en péptidos, evitando redundancias que puedan generar respuestas inmunogénicas no deseadas.\n",
    "\n",
    "4. Análisis evolutivo\n",
    "\n",
    "- Ayuda a estudiar la conservación de secuencias en proteínas homólogas y en el desarrollo de alineamientos múltiples.\n",
    "\n",
    "El RRI es una métrica valiosa para analizar la redundancia de residuos en secuencias peptídicas. Su aplicación en bioinformática estructural, diseño de péptidos y análisis evolutivo permite mejorar la comprensión de la relación entre la secuencia y la función de las proteínas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RRI_A</th>\n",
       "      <th>RRI_C</th>\n",
       "      <th>RRI_D</th>\n",
       "      <th>RRI_E</th>\n",
       "      <th>RRI_F</th>\n",
       "      <th>RRI_G</th>\n",
       "      <th>RRI_H</th>\n",
       "      <th>RRI_I</th>\n",
       "      <th>RRI_K</th>\n",
       "      <th>RRI_L</th>\n",
       "      <th>RRI_M</th>\n",
       "      <th>RRI_N</th>\n",
       "      <th>RRI_P</th>\n",
       "      <th>RRI_Q</th>\n",
       "      <th>RRI_R</th>\n",
       "      <th>RRI_S</th>\n",
       "      <th>RRI_T</th>\n",
       "      <th>RRI_V</th>\n",
       "      <th>RRI_W</th>\n",
       "      <th>RRI_Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.67</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.33</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.57</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.57</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.57</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2700</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.33</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2704 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RRI_A  RRI_C  RRI_D  RRI_E  RRI_F  RRI_G  RRI_H  RRI_I  RRI_K  RRI_L  \\\n",
       "0       1.0    0.0    0.0    0.0    0.0    1.0    1.0    1.5   1.67    1.0   \n",
       "1       1.0    0.0    0.0    0.0    1.0    1.0    1.0    1.0   1.57    1.0   \n",
       "2       1.0    0.0    0.0    0.0    1.0    1.0    1.0    1.0   1.57    1.0   \n",
       "3       1.0    0.0    0.0    0.0    1.0    0.0    1.0    1.0   1.57    1.0   \n",
       "4       1.0    0.0    1.0    0.0    1.0    1.0    0.0    1.0   2.00    1.0   \n",
       "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "2699    1.0    1.0    0.0    0.0    0.0    1.0    1.0    0.0   0.00    0.0   \n",
       "2700    1.0    1.0    0.0    0.0    0.0    1.0    1.0    0.0   0.00    0.0   \n",
       "2701    0.0    1.0    0.0    0.0    1.0    1.0    0.0    0.0   2.00    1.0   \n",
       "2702    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0   1.00    0.0   \n",
       "2703    1.0    0.0    0.0    0.0    1.0    1.4    0.0    1.0   1.00    1.0   \n",
       "\n",
       "      RRI_M  RRI_N  RRI_P  RRI_Q  RRI_R  RRI_S  RRI_T  RRI_V  RRI_W  RRI_Y  \n",
       "0       0.0    0.0    1.0    0.0   1.33    0.0    1.0    1.0    0.0    1.0  \n",
       "1       0.0    0.0    0.0    0.0   0.00    1.0    0.0    0.0    1.0    0.0  \n",
       "2       0.0    0.0    1.0    0.0   0.00    1.0    0.0    0.0    1.0    0.0  \n",
       "3       0.0    0.0    1.0    0.0   0.00    1.0    0.0    0.0    1.0    0.0  \n",
       "4       0.0    0.0    0.0    0.0   0.00    1.0    0.0    1.0    0.0    0.0  \n",
       "...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  \n",
       "2699    0.0    0.0    0.0    0.0   8.00    1.0    0.0    0.0    1.0    0.0  \n",
       "2700    0.0    0.0    0.0    0.0   1.40    0.0    0.0    0.0    1.0    0.0  \n",
       "2701    0.0    0.0    2.0    1.0   2.33    0.0    0.0    0.0    0.0    0.0  \n",
       "2702    0.0    0.0    0.0    0.0   9.00    0.0    0.0    0.0    0.0    0.0  \n",
       "2703    0.0    0.0    1.0    2.0   1.00    1.0    2.0    1.0    1.0    0.0  \n",
       "\n",
       "[2704 rows x 20 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def rri(input):\n",
    "  a = input.rstrip('txt')\n",
    "  output = a + 'rri.csv'\n",
    "  df_out = rri_wp(input, output)\n",
    "  df_in = pd.read_csv(output)\n",
    "  return df_in\n",
    "\n",
    "feature_rri = rri(cd_hit_path)\n",
    "feature_rri.to_csv('../data/processed/feature_rri.csv', index = False)\n",
    "feature_rri"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusiones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Referencias\n",
    "\n",
    "1. INTERNATIONAL UNION OF PURE AND APPLIED CHEMISTRY (IUPAC). https://iupac.qmul.ac.uk/AminoAcid/A2021.html#AA20\n",
    "2. Fu, L., Niu, B., Zhu, Z., Wu, S., & Li, W. (2012). CD-HIT: accelerated for clustering the next-generation sequencing data. Bioinformatics (Oxford, England), 28(23), 3150–3152. https://doi.org/10.1093/bioinformatics/bts565\n",
    "3. Raghava, S. P. a. D. a. P. D. K. N. S. G. M. R. K. P. a. V. P. a. L. H. K. S. J. P. G. (n.d.). Pfeature: A web server for computing protein and peptide features. https://webs.iiitd.edu.in/raghava/pfeature/\n",
    "4. Dubchak I, et al. Proceedings of the National Academy of Sciences of the United States of America"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
